// @makarandkate/invoice-templates/src/models/expense.model.ts
import { ExpenseItem } from './expense-item.model';
import Party from './party.model';
import { MoneyOut } from './money-out.model';
import { Arrays } from '../utils/ReferenceData';
import { PurchaseItem } from './purchase-item.model';

export class Expense {
    id: number;
    localId: string;
    parentId: number;
    invoiceNo: string;
    billNo: string;
    partyId: number;
    partyIdLocal: string;
    deliveryState: keyof typeof Arrays.states;
    expensePic:string;
    senderState: string;
    invoiceDateStamp: number;
    billDateStamp?: number;
    dueDateStamp: number;
    isRecurring: boolean;
    recurringDays: number;
    eWayBillNo: string;
    eWayBillDate: number;
    trmAndCnd: boolean;
    description: string;
    itemAmount: number;
    additionalAmount: number;
    discountAmount: number;
    tdsAmount: number;
    tdsPercentage:number;
    deliveryDate:number;
    gstAmount: number;
    cessAmount: number;
    totalAmount: number;
    paidStamp: number;
    cancelStamp: number;
    createdStamp: number;
    gstEnabled: boolean;
    template: string;
    transporterName: string;
    vehicleNumber: string;
    deliveryLocation: string;
    color: string;
    url: string;
    invoiceReadStamp: string;
    openCount: number;
    additinalAttr: string;
    pic: string;
    acceptStamp: number;
    acceptSignature: number;
    updatedStamp: number;
    expenseItems: ExpenseItem[];
    //Extras
    partyData?: Party;
    note: string;
    category?: string;
     //have to emove keys beofre call
    indexDbKey?:string;
    isDraft?:boolean;
    moneyOut?: MoneyOut;
    moneyOutId?: number;
    moneyOutIdLocal?: string;
    purchasePic?:string;
    purchaseItems?: PurchaseItem[];
    partyName?:string;
    isDuplicateBillEntry? : boolean;
    isDuplicateInvNo? : boolean;
}
