// @makarandkate/invoice-templates/src/models/MoneyOutBillPrint.model.ts

export class MoneyOutBillPrint {

    user: {
        isPro?: number;
        registrationStamp?: number;
    };

    profile: {
        addressProvience?: string;
        legalName?: string;
        addressLine1?: string;
        addressLine2?: string;
        addressPostalCode?: string;
        contactPersonPhone?: string;
        contactPersonEmail?: string;
        gstin?: string;
        fssaiNumber?: string;
        licenseNumber?: string;
        bankAccountNo?: string;
        bankName?: string;
        bankIFSC?: string;
        bankAccountType?: string;
        bankUPI?: string;
        additionalInputFieldTitle?: string;
        additionalInputFieldValue?: string;
        additionalDateFieldValue?: string; // Not Avail
        additionalDateFieldTitle?: string; // Not Avail
        signatureLink?: string;
        logoLink?: string;
    };

    bill?: {
        partyData: {
            ledgerCredit?: number;
            profileData: {
                legalName?: string;
                gstNumber?: string;
                contactPersonName?: string;
                contactPersonPhone?: string;
                contactPersonEmail?: string;
                addressLine1?: string;
                addressState?: string;
                addressPincode?: string;
                addressOneLine1?: string;
                addressOneState?: string;
                addressOnePincode?: string;
                udf1T?: string;
                udf1V?: string;
                udf2T?: string;
                udf2V?: string;
                udf3T?: string;
                udf3V?: string;
                additionalInputFieldTitle?: string;
                additionalInputFieldValue?: string;
                additionalDateFieldValue?: string; // Not Avail
                additionalDateFieldTitle?: string; // Not Avail
            };
        };
        txnMode?: string;
        txnRef?: string;
        isPartyDeleted?:boolean;
        billNo?: string;
        billDateStamp?: number;
        createdStamp?: number;
        color: string;
        note?: string;
        billTitle?: string;
        tnc?: string;
        totalAmount?:number;
    };

    settings?: {
        showSignatureNotRequiredText?: boolean;
        showSignatureInTemplates?: boolean;
        include3InchOnlineDukanLink?: boolean;
        iSetPrintTimeOnBill?: boolean;
    };

    onlineDukan?: {
        domain: string;
    }

    extra?: {
        thermalTncTemplate?: string;
    }

    qrPayLink?: string;
    addressTo?: string;
    pageWidth?: number;

}
