// @makarandkate/invoice-templates/src/models/print.model.ts
import Party from './party.model';
import { SenderProfile } from './sender-profile';
import { Invoice } from './invoice.model';
import { MoneyIn } from './money-in.model';
import { ProfileData } from './profile-data.model';
import { Purchase } from './purchase.model';
import { MoneyOut } from './money-out.model';
import { User } from './user.model';

export class InvoicePrint extends Invoice {
    user: SenderProfile; // user?.profileData => profileData
    fbUser?:User; // fbUser => user
    partyInfo: Party; // partyInfo => invoice.partyData
    moneyInInfo?: MoneyIn; // moneyInInfo => invoice.moneyInOut || moneyInOut
    moneyOutInfo?: MoneyOut; // moneyOutInfo => invoice.moneyInOut || moneyInOut
    signature?: string; // signature => profileData.signature
    logo?: string; // logo => profileData.logo
    payLink?: string; // Removed
    qrPayLink?:string; // No change
    isEstimate?:boolean|undefined; // No change
    isPurchase?:boolean|undefined; // No change
    isSaleReturn?:boolean|undefined; // No change
    isPurchaseReturn?:boolean|undefined; // No change
    isMoneyIn?:boolean|undefined; // No change
    isMoneyOut?:boolean|undefined; // No change
    moneyOut?:MoneyOut; // Removed
    addressTo?: string; // No change
    showSignatureNotRequiredText?:boolean; // showSignatureNotRequiredText => settings.showSignatureNotRequiredText
    showSignatureInTemplates?: boolean; // showSignatureInTemplates => settings.showSignatureInTemplates
    makeItemNameBold?:boolean; // makeItemNameBold => settings.makeItemNameBold
    showPreviousAmount?:boolean; // showPreviousAmount => settings.showPreviousAmount
    showTaxAndDiscAmt?:boolean; // showTaxAndDiscAmt => settings.showTaxAndDiscAmt
    roundOffValue?:number; // roundOffValue => invoice.roundOffValue
    pageWidth?:number; // No change
    refInvoiceNumber?:string; // refInvoiceNumber => invoice.refInvoiceNumber
    // other data wrap to invoice object
}

export class MoneyInPrint extends MoneyIn {
    user: SenderProfile;
    partyInfo: Party;
    signature?: string;
    logo?: string;
}

export class MoneyOutPrint extends MoneyOut {
    user: SenderProfile;
    partyInfo: Party;
    signature?: string;
    logo?: string;
}

export class PurchasePrint extends Purchase {
    user: SenderProfile;
    partyInfo: ProfileData;
    items?: any;    
    logo?: string;
    isPurchase?:boolean|undefined;
    isSaleReturn?:boolean|undefined;
    moneyOutInfo: MoneyOut;
    signature?: string;
}
