// @makarandkate/invoice-templates/src/models/profile-data.model.ts
// export class Profiledata {
//     id: number;
//     brandName: string;
//     legalName: string;
//     entityType: string;
//     businessType: string;
//     contactPersonName: string;
//     contactPersonPhone: string;
//     contactPersonEmail: string;
//     landLineNumber: string;
//     pan: string;
//     isGstEnabled: boolean;
//     gstNumber: string;
//     gstType: string;
//     addressLine1: string;
//     addressCity: string;
//     addressState: string;
//     logo: string;
//     tan: string;
//     bankAccountNo: string;
//     ifscCode: string;
//     accountType: string;
//     upi: string;
//     signature: string;
//     termsCondition: string;
//     isPrimary: boolean;
//     createdStamp: number;
//     updatedStamp: number;
// }

import { Arrays } from '../utils/ReferenceData';
export class ProfileData {
    id?: number;
    localId?: string;
    brandName?: string;
    profileName?:string;
    legalName?: string;
    category?: string;
    entityType?: string;
    businessType?: string;
    contactPersonName?: string;
    contactPersonPhone?: string;
    contactPersonEmail?: string;
    landLineNumber?: string;
    pan?: string;
    isGstEnabled?: boolean;
    gstNumber?: string;
    gstType?: string;
    addressLine1?: string;
    addressCity?: string;
    addressState?: keyof typeof Arrays.states;
    addressPincode?: number;
    addressOneLine1?: string;
    addressOneCity?: string;
    addressOneState?: keyof typeof Arrays.states;
    addressOnePincode?: number;
    logo?: string;
    tan?: string;
    language?: string;
    template?: string;
    bankAccountNo?: string;
    bankName?: string;
    ifscCode?: string;
    accountType?: string;
    upi?: string;
    signature?: string;
    termsCondition?: string;
    thermalTncTemplate?: string;
    isPrimary?: boolean;
    createdStamp?: number;
    updatedStamp?: number;
    geoLocation?: string;
    invoiceTitle?: string;
    pdfViewFooter?: number;
    invoiceTemplate?: string;
    invoiceTempColor?: string;
    creditLimit?:number;
    monthlyAlertDate?:number;
    udf1T?: string;
    tags?:{items?:string[],parties?:string[]};
    udf1V?: string;
    udf2T?: string;
    udf2V?: string;
    udf3T?: string;
    udf3V?: string;
    onlineDukanColor?: string; 
    zeroInventoryItem?: number;
    wholeSalerRate?: number;
    reduceQty?: number;
    smsAlertPref?: any;
    odEnableCod?: number;
    odEnableUpi?: number;
    additionalData?: {
        itemBeepSound?:1|0,
        autoFillAmtReceived?:1|0,
        defaultCashInvoice?:1|0,
        smsToBusinessOwner?:1|0,
        invoiceSmsToUser?:1|0,
        include3InchOnlineDukanLink?:0|1
    };
    additionalItemFormFields?:{inputFieldTitle?:string,inputFieldValue?:string,dateFieldTitle?:string,dateFieldValue?:string};
    additionalInputFieldValue?:string;
    additionalDateFieldValue?: string;
    additionalInputFieldTitle?:string;
    additionalDateFieldTitle?: string;
}
