// @makarandkate/invoice-templates/src/models/purchase-return.model.ts
import { InvoiceItem } from './invoice-item.model'
import { Party } from './Party';
import { MoneyIn } from './money-in.model';
import { Arrays } from '../utils/ReferenceData';

export class PurchaseReturn {
    id?:number;
    localId?:string;
    parentId?: number;
    invoiceNo?: string;
    billNo?: string;
    partyId?: number;
    partyIdLocal: string;
    deliveryState?: keyof typeof Arrays.states;;
    senderState?: string;
    invoiceDateStamp: number;
    billDateStamp?: number;
    dueDateStamp: number;
    isRecurring: boolean;
    recurringDays: number;
    purOrderNo?: string;
    eWayBillNo: string;
    trmAndCnd: boolean;
    description: string;
    itemAmount: number;
    additionalAmount: number;
    discountAmount: number;
    tdsAmount: number;
    tdsPercentage?: number;
    gstAmount: number;
    cessAmount: number;
    totalAmount: number;
    paidStamp: number;
    cancelStamp?: number;
    createdStamp?: number;
    gstEnabled: boolean;
    template: string;
    transporterName: string;
    transDistance?: string;
    vehicleNumber: string;
    deliveryLocation: string;
    deliveryDate?: number;
    color: string;
    url: string;
    invoiceReadStamp: string;
    openCount: number;
    additinalAttr?: string;
    pic: string;
    acceptStamp?: number;
    acceptSignature?: number;
    updatedStamp?: number;
    moneyinId?: number;
    moneyinIdLocal?: string;
    moneyIn?: MoneyIn;
    invoiceItems?: InvoiceItem[];
    partyName?: string;
    partyData?: Party;
    eWayBillDate: number;
    note: string;
    invoiceId?: number;
    invConvertedStamp?: number;
    estId?: number;
    invoiceTitle?: string;
    textColor?:string;
    tnc?:string;
    isPro?:any;
    shareHashUrl?:string;
    amount?:number;
    //remvoe before making ajax calls
    indexDbKey?:string;
    poId? : number;
    seenStamp?: number;
    isDraft?:boolean;
    error?:any;
    flatDiscount?: number;
    flatDiscountTimeStamp?: number;
    isDuplicateBillEntry? : boolean;
    isDuplicateInvNo? : boolean;
}

export class PurchaseReturnFormSubmitType{
    gstAmount: number;
    cessAmount:  number;
    totalAmount:  number;
    discountAmount:  number;
    dueDateStamp:  number;
    additionalAmount:  number;
    eWayBillNo:  string;
    eWayBillDate:  number;
    purOrderNo: string;
    transporterName: string;
    vehicleNumber:string;
    deliveryLocation: string;
    deliveryDate:  number;
    transDistance: string;
    tnc: string;
    pic: string;
    invoiceNo: string;
    invoiceDateStamp:  number;
    flatDiscount: number;
    invoiceTitle: string;
    note: string;
    partyId:  string;
    deliveryState?:keyof typeof Arrays.states;
    invoiceItems:InvoiceItem[];
    template: string;
    color: string;
    partyName: string;
    moneyIn: MoneyIn
    textColor: string;
    moneyInAmount: number;
  }
