// @makarandkate/invoice-templates/src/models/sale-return.model.ts
import { PurchaseItem } from './purchase-item.model';
import Party from './party.model';
import { MoneyOut } from './money-out.model';
import { Arrays } from '../utils/ReferenceData';

export class SaleReturn {
    id: number;
    localId: string;
    parentId: number;
    invoiceNo: string;
    billNo?: string;
    partyId: number;
    partyIdLocal: string;
    deliveryState?:keyof typeof Arrays.states;
    purchasePic?:string;
    senderState: string;
    invoiceDateStamp: number;
    billDateStamp?: number;
    dueDateStamp: number;
    isRecurring: boolean;
    recurringDays: number;
    eWayBillNo: string;
    eWayBillDate: number;
    trmAndCnd: boolean;
    description: string;
    itemAmount: number;
    additionalAmount: number;
    discountAmount: number;
    tdsAmount: number;
    tdsPercentage:number;
    deliveryDate:number;
    gstAmount: number;
    cessAmount: number;
    totalAmount: number;
    paidStamp: number;
    cancelStamp: number;
    createdStamp: number;
    gstEnabled: boolean;
    template: string;
    transporterName: string;
    vehicleNumber: string;
    deliveryLocation: string;
    color: string;
    url: string;
    invoiceReadStamp: string;
    openCount: number;
    additinalAttr: string;
    pic: string;
    acceptStamp: number;
    acceptSignature: number;
    updatedStamp: number;
    purchaseItems?: PurchaseItem[];
    saleReturnItems?: PurchaseItem[];
    partyName?:string;
    //Extras
    partyData?: Party;
    note: string;
    category?: string;
    indexDbKey?: string;
    isDraft?: boolean;
    moneyOut?: MoneyOut;
    moneyOutId?: number;
    moneyOutIdLocal?: string;
    receiptNo?: string;
    isDuplicateBillEntry?: boolean;
    isDuplicateInvNo?: boolean;
    amount?: number;
    dateStamp?: number
}
