// @makarandkate/invoice-templates/lib/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HtmlTemplates = void 0;
var MoneyInBillPrint_model_1 = require("./models/MoneyInBillPrint.model");
var gst1_A5_1 = require("./templates/gst1-A5");
var gst2_A5_1 = require("./templates/gst2-A5");
var letter_head_1 = require("./templates/letter-head");
var modern_A5_1 = require("./templates/modern-A5");
var tallyGST_1 = require("./templates/tallyGST");
var tallyHSN_1 = require("./templates/tallyHSN");
var temp1_1 = require("./templates/temp1");
var temp1_2 = require("./templatesV2/temp1");
var temp1_money_in_1 = require("./templatesV2/temp1-money-in");
var temp1_money_out_1 = require("./templatesV2/temp1-money-out");
var temp1_purchase_1 = require("./templatesV2/temp1-purchase");
var temp1_expense_1 = require("./templatesV2/temp1-expense");
var temp1_estimate_1 = require("./templatesV2/temp1-estimate");
var temp17_1 = require("./templatesV2/temp17");
var temp17_estimate_1 = require("./templatesV2/temp17-estimate");
var temp17_money_in_1 = require("./templatesV2/temp17-money-in");
var temp17_money_out_1 = require("./templatesV2/temp17-money-out");
var temp17_purchase_1 = require("./templatesV2/temp17-purchase");
var temp17_expense_1 = require("./templatesV2/temp17-expense");
var temp18_1 = require("./templatesV2/temp18");
var temp18_money_in_1 = require("./templatesV2/temp18-money-in");
var temp18_money_out_1 = require("./templatesV2/temp18-money-out");
var temp18_purchase_1 = require("./templatesV2/temp18-purchase");
var temp18_expense_1 = require("./templatesV2/temp18-expense");
var temp18_estimate_1 = require("./templatesV2/temp18-estimate");
var temp19_1 = require("./templatesV2/temp19");
var temp19_money_in_1 = require("./templatesV2/temp19-money-in");
var temp19_money_out_1 = require("./templatesV2/temp19-money-out");
var temp19_purchase_1 = require("./templatesV2/temp19-purchase");
var temp19_expense_1 = require("./templatesV2/temp19-expense");
var temp19_estimate_1 = require("./templatesV2/temp19-estimate");
var temp20_1 = require("./templatesV2/temp20");
var temp20_money_in_1 = require("./templatesV2/temp20-money-in");
var temp20_money_out_1 = require("./templatesV2/temp20-money-out");
var temp20_purchase_1 = require("./templatesV2/temp20-purchase");
var temp20_expense_1 = require("./templatesV2/temp20-expense");
var temp20_estimate_1 = require("./templatesV2/temp20-estimate");
var temp17_2 = require("./templates/temp17");
var temp2_1 = require("./templates/temp2");
var temp3_1 = require("./templates/temp3");
var temp18_2 = require("./templates/temp18");
var BillPrint_model_1 = require("./models/BillPrint.model");
var SaleBillPrint_model_1 = require("./models/SaleBillPrint.model");
var PurchaseBillPrint_model_1 = require("./models/PurchaseBillPrint.model");
var ExpenseBillPrint_model_1 = require("./models/ExpenseBillPrint.model");
var MoneyOutBillPrint_model_1 = require("./models/MoneyOutBillPrint.model");
var EstimateBillPrint_model_1 = require("./models/EstimateBillPrint.model");
var HtmlTemplates = /** @class */ (function () {
    function HtmlTemplates() {
    }
    HtmlTemplates.newLineToBr = function (inputString) {
        return ((inputString === null || inputString === void 0 ? void 0 : inputString.length) > 0) ? inputString.replace(/\n/g, '<br/>') : "";
    };
    HtmlTemplates.getTemplateHtml = function (data, invoicePref) {
        return __awaiter(this, void 0, void 0, function () {
            var templateObj, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        templateObj = {
                            temp6: function (x) { return new temp1_1.Temp1(x).main(); },
                            temp2: function (x) { return new temp2_1.Temp2(x).main(); },
                            temp3: function (x) { return new temp3_1.Temp3(x).main(); },
                            temp9: function (x) { return new modern_A5_1.ModernA5(x).main(); },
                            temp10: function (x) { return new gst1_A5_1.Gst1A5(x).main(); },
                            temp11: function (x) { return new gst2_A5_1.Gst2A5(x).main(); },
                            temp12: function (x) { return new tallyHSN_1.TallyHSN(x).main(); },
                            temp13: function (x) { return new tallyGST_1.TallyGST(x).main(); },
                            temp14: function (x) { return new letter_head_1.LetterHead(x).main(); },
                            temp17: function (x) { return new temp17_2.Temp17(x).main(); },
                            temp18: function (x) { return new temp18_2.Temp18(x).main(); },
                        };
                        if (!templateObj[invoicePref]) return [3 /*break*/, 2];
                        return [4 /*yield*/, templateObj[invoicePref](data)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2: return [4 /*yield*/, templateObj.temp6(data)];
                    case 3: return [2 /*return*/, _a.sent()];
                    case 4:
                        error_1 = _a.sent();
                        return [3 /*break*/, 5];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    HtmlTemplates.generateBillObject = function (args) {
        var _this = this;
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36, _37, _38, _39, _40, _41, _42, _43, _44, _45, _46, _47, _48, _49;
        var billPrint = new BillPrint_model_1.BillPrint();
        billPrint.user = args.user;
        billPrint.profile = {
            addressProvience: (_a = args.profile) === null || _a === void 0 ? void 0 : _a.addressProvience,
            legalName: (_b = args.profile) === null || _b === void 0 ? void 0 : _b.legalName,
            addressLine1: (((_c = args.profile) === null || _c === void 0 ? void 0 : _c.addressLine1) || '') + (((_d = args.profile) === null || _d === void 0 ? void 0 : _d.addressLine2) || ''),
            addressPostalCode: (_e = args.profile) === null || _e === void 0 ? void 0 : _e.addressPostalCode,
            contactPersonPhone: (_f = args.profile) === null || _f === void 0 ? void 0 : _f.contactPersonPhone,
            contactPersonEmail: (_g = args.profile) === null || _g === void 0 ? void 0 : _g.contactPersonEmail,
            gstin: (_h = args.profile) === null || _h === void 0 ? void 0 : _h.gstin,
            fssaiNumber: (_j = args.profile) === null || _j === void 0 ? void 0 : _j.fssaiNumber,
            licenseNumber: (_k = args.profile) === null || _k === void 0 ? void 0 : _k.licenseNumber,
            bankAccountNo: (_l = args.profile) === null || _l === void 0 ? void 0 : _l.bankAccountNo,
            bankName: (_m = args.profile) === null || _m === void 0 ? void 0 : _m.bankName,
            bankIFSC: (_o = args.profile) === null || _o === void 0 ? void 0 : _o.bankIFSC,
            bankAccountType: (_p = args.profile) === null || _p === void 0 ? void 0 : _p.bankAccountType,
            bankUPI: (_q = args.profile) === null || _q === void 0 ? void 0 : _q.bankUPI,
            additionalInputFieldTitle: (_r = args.profile) === null || _r === void 0 ? void 0 : _r.additionalInputFieldTitle,
            additionalInputFieldValue: (_s = args.profile) === null || _s === void 0 ? void 0 : _s.additionalInputFieldValue,
            additionalDateFieldValue: (_t = args.profile) === null || _t === void 0 ? void 0 : _t.additionalDateFieldValue,
            additionalDateFieldTitle: (_u = args.profile) === null || _u === void 0 ? void 0 : _u.additionalDateFieldTitle,
            signatureLink: ((_v = args.signature) === null || _v === void 0 ? void 0 : _v.imageBase64) || '',
            logoLink: ((_w = args.profile) === null || _w === void 0 ? void 0 : _w.pSetLogoPrintStatus) ? ((_x = args.logo) === null || _x === void 0 ? void 0 : _x.imageBase64) || '' : '',
        };
        billPrint.bill = {
            partyData: {
                ledgerCredit: (_y = args.party) === null || _y === void 0 ? void 0 : _y.credit,
                profileData: {
                    legalName: (_z = args.party) === null || _z === void 0 ? void 0 : _z.businessName,
                    gstNumber: (_0 = args.party) === null || _0 === void 0 ? void 0 : _0.gstin,
                    contactPersonName: (_1 = args.party) === null || _1 === void 0 ? void 0 : _1.name,
                    contactPersonPhone: (_2 = args.party) === null || _2 === void 0 ? void 0 : _2.phone,
                    contactPersonEmail: (_3 = args.party) === null || _3 === void 0 ? void 0 : _3.email,
                    addressLine1: (_4 = args.party) === null || _4 === void 0 ? void 0 : _4.billingAddress,
                    addressState: (_5 = args.party) === null || _5 === void 0 ? void 0 : _5.billingProvience,
                    addressPincode: (_6 = args.party) === null || _6 === void 0 ? void 0 : _6.billingPostalCode,
                    addressOneLine1: (_7 = args.party) === null || _7 === void 0 ? void 0 : _7.deliveryAddress,
                    addressOneState: (_8 = args.party) === null || _8 === void 0 ? void 0 : _8.deliveryProvience,
                    addressOnePincode: (_9 = args.party) === null || _9 === void 0 ? void 0 : _9.deliveryPostalCode,
                    udf1T: '',
                    udf1V: '',
                    udf2T: '',
                    udf2V: '',
                    udf3T: '',
                    udf3V: '',
                    additionalInputFieldTitle: '',
                    additionalInputFieldValue: '',
                    additionalDateFieldValue: '',
                    additionalDateFieldTitle: '',
                },
            },
            moneyInOut: {
                billNo: (_10 = args.moneyIn) === null || _10 === void 0 ? void 0 : _10.billNo,
                dateStamp: (_11 = args.moneyIn) === null || _11 === void 0 ? void 0 : _11.billDateStamp,
                amount: ((_12 = args.sale) === null || _12 === void 0 ? void 0 : _12.amountReceived) || ((_13 = args.moneyIn) === null || _13 === void 0 ? void 0 : _13.totalAmount),
                txnMode: (_14 = args.moneyIn) === null || _14 === void 0 ? void 0 : _14.paymentMode,
                txnRef: (_15 = args.moneyIn) === null || _15 === void 0 ? void 0 : _15.paymentId,
            },
            billItems: (_16 = args.sale.billItems) === null || _16 === void 0 ? void 0 : _16.map(function (billItem) {
                var _a, _b, _c, _d, _e, _f, _g;
                return {
                    itemDes: _this.newLineToBr((billItem === null || billItem === void 0 ? void 0 : billItem.billNote) || ''),
                    itemSrl: '',
                    hsn: ((_a = billItem.item) === null || _a === void 0 ? void 0 : _a.hsn) || '',
                    price: _this.capFractionsToTwo(billItem.price),
                    basePrice: _this.capFractionsToTwo(billItem.basePrice),
                    incTax: billItem.incTax,
                    taxPercentage: _this.capFractionsToTwo(billItem.taxPercentage),
                    cessPercentage: _this.capFractionsToTwo(billItem.cessPercentage),
                    unit: billItem.unit,
                    quantity: billItem.quantity,
                    discountPercentage: _this.capFractionsToTwo(billItem.discountPercentage),
                    discount: _this.capFractionsToTwo(billItem.discount),
                    unitGstAmount: _this.capFractionsToTwo(billItem.unitGstAmount),
                    unitCessAmount: _this.capFractionsToTwo(billItem.unitCessAmount),
                    unitDiscountAmount: _this.capFractionsToTwo(billItem.unitDiscountAmount),
                    unitTaxAmount: _this.capFractionsToTwo(billItem.unitTaxAmount),
                    itemTotalGstAmount: _this.capFractionsToTwo(billItem.itemTotalGstAmount),
                    itemTotalCessAmount: _this.capFractionsToTwo(billItem.itemTotalCessAmount),
                    itemTotalTaxAmount: _this.capFractionsToTwo(billItem.itemTotalTaxAmount),
                    effectiveTaxPercentage: _this.capFractionsToTwo(billItem.effectiveTaxPercentage),
                    billCashDiscountPercentage: _this.capFractionsToTwo(billItem.billCashDiscountPercentage),
                    total: _this.capFractionsToTwo(billItem.wcdTotal),
                    createdStamp: (_b = billItem.item) === null || _b === void 0 ? void 0 : _b.createdStamp,
                    item: {
                        itemName: (_c = billItem.item) === null || _c === void 0 ? void 0 : _c.itemName,
                        hsn: (_d = billItem.item) === null || _d === void 0 ? void 0 : _d.hsn,
                        mrp: (_e = billItem.item) === null || _e === void 0 ? void 0 : _e.mrp,
                        expiryDate: (_f = billItem.item) === null || _f === void 0 ? void 0 : _f.expiryDate,
                    },
                    additionalInputFieldValue: '',
                    additionalDateFieldValue: 0,
                    additionalInputFieldTitle: '',
                    additionalDateFieldTitle: '',
                    mrp: ((_g = billItem.item) === null || _g === void 0 ? void 0 : _g.mrp) || null,
                    variant: '',
                };
            }),
            roundOffValue: this.capFractionsToTwo(args.sale.roundOffValue) || 0,
            billNo: args.sale.billNo,
            deliveryState: args.sale.party.deliveryProvience,
            billDateStamp: args.sale.billDateStamp,
            dueDateStamp: args.sale.dueDateStamp,
            purOrderNo: (_17 = args.sale.transportDetail) === null || _17 === void 0 ? void 0 : _17.purOrderNo,
            eWayBillNo: (_18 = args.sale.transportDetail) === null || _18 === void 0 ? void 0 : _18.eWayBillNo,
            additionalAmount: this.capFractionsToTwo((_19 = args.sale.transportDetail) === null || _19 === void 0 ? void 0 : _19.additionalAmount),
            discountAmount: this.capFractionsToTwo(args.sale.discountAmount),
            gstAmount: this.capFractionsToTwo(args.sale.gstAmount),
            cessAmount: this.capFractionsToTwo(args.sale.cessAmount),
            totalAmount: this.capFractionsToTwo(args.sale.totalAmount),
            createdStamp: args.sale.createdStamp,
            transporterName: (_20 = args.sale.transportDetail) === null || _20 === void 0 ? void 0 : _20.transporterName,
            vehicleNumber: (_21 = args.sale.transportDetail) === null || _21 === void 0 ? void 0 : _21.vehicleNumber,
            deliveryDate: (_22 = args.sale.transportDetail) === null || _22 === void 0 ? void 0 : _22.deliveryDate,
            color: '#3498db',
            eWayBillDate: (_23 = args.sale.transportDetail) === null || _23 === void 0 ? void 0 : _23.eWayBillDate,
            note: args.sale.note,
            billTitle: '',
            tnc: ((_24 = args.profile) === null || _24 === void 0 ? void 0 : _24.pSetTermsAndConditions) || null,
            cashDiscount: this.capFractionsToTwo(args.sale.cashDiscount),
            amountReceived: this.capFractionsToTwo(args.sale.amountReceived),
            cashDiscountPercentage: this.capFractionsToTwo(args.sale.cashDiscountPercentage),
            challanNo: ((_25 = args.sale.transportDetail) === null || _25 === void 0 ? void 0 : _25.challanNo) || null,
            totalSaving: this.capFractionsToTwo(args.sale.totalSaving),
            subTotalAmount: this.capFractionsToTwo(args.sale.subTotalAmount),
            partyPreviousBalance: this.capFractionsToTwo(args.sale.partyPreviousBalance) || null,
            billCompleteStamp: args.sale.billCompleteStamp,
            serviceChargePercentage: this.capFractionsToTwo(args.sale.serviceChargePercentage),
            serviceChargeAmount: this.capFractionsToTwo(args.sale.serviceChargeAmount),
            deliveryProvience: (_26 = args === null || args === void 0 ? void 0 : args.sale) === null || _26 === void 0 ? void 0 : _26.deliveryProvience,
        };
        if ((_27 = args.secondaryParty) === null || _27 === void 0 ? void 0 : _27._localUUID) {
            billPrint.bill.secondaryPartyData = {
                ledgerCredit: (_28 = args.secondaryParty) === null || _28 === void 0 ? void 0 : _28.credit,
                profileData: {
                    legalName: (_29 = args.secondaryParty) === null || _29 === void 0 ? void 0 : _29.businessName,
                    gstNumber: (_30 = args.secondaryParty) === null || _30 === void 0 ? void 0 : _30.gstin,
                    contactPersonName: (_31 = args.secondaryParty) === null || _31 === void 0 ? void 0 : _31.name,
                    contactPersonPhone: (_32 = args.secondaryParty) === null || _32 === void 0 ? void 0 : _32.phone,
                    contactPersonEmail: (_33 = args.secondaryParty) === null || _33 === void 0 ? void 0 : _33.email,
                    addressLine1: (_34 = args.secondaryParty) === null || _34 === void 0 ? void 0 : _34.billingAddress,
                    addressState: (_35 = args.secondaryParty) === null || _35 === void 0 ? void 0 : _35.billingProvience,
                    addressPincode: (_36 = args.secondaryParty) === null || _36 === void 0 ? void 0 : _36.billingPostalCode,
                    addressOneLine1: (_37 = args.secondaryParty) === null || _37 === void 0 ? void 0 : _37.deliveryAddress,
                    addressOneState: (_38 = args.secondaryParty) === null || _38 === void 0 ? void 0 : _38.deliveryProvience,
                    addressOnePincode: (_39 = args.secondaryParty) === null || _39 === void 0 ? void 0 : _39.deliveryPostalCode,
                    udf1T: '',
                    udf1V: '',
                    udf2T: '',
                    udf2V: '',
                    udf3T: '',
                    udf3V: '',
                    additionalInputFieldTitle: '',
                    additionalInputFieldValue: '',
                    additionalDateFieldValue: '',
                    additionalDateFieldTitle: '',
                },
            };
        }
        // set moneyInOut
        billPrint.moneyInOut = null;
        // set bill type flag
        billPrint.isEstimate = false;
        billPrint.isPurchase = false;
        billPrint.isSaleReturn = false;
        billPrint.isPurchaseReturn = false;
        billPrint.isMoneyIn = false;
        billPrint.isMoneyOut = false;
        // set settings - TODO All
        billPrint.settings = {
            showSignatureNotRequiredText: false,
            showSignatureInTemplates: true,
            makeItemNameBold: false,
            showPreviousAmount: false,
            showTaxAndDiscAmt: (_40 = args.profile) === null || _40 === void 0 ? void 0 : _40.pSetTaxMrpHsnPrintStatus,
            billSavingsAmount: true,
        };
        // set extra - TODO All
        billPrint.extra = {
            thermalTncTemplate: ((_41 = args.profile) === null || _41 === void 0 ? void 0 : _41.pSetTermsAndConditions) || null,
        };
        billPrint.qrPayLink = '';
        billPrint.addressTo = '';
        billPrint.pageWidth = null;
        if (((_42 = args.profile) === null || _42 === void 0 ? void 0 : _42.bankUPI) && ((_43 = args.profile) === null || _43 === void 0 ? void 0 : _43.pSetQrPrintStatus)) {
            var receivedAmount_1 = 0;
            var balancedAmount = 0;
            (_44 = args.sale) === null || _44 === void 0 ? void 0 : _44.moneyIns.forEach(function (moneyIn) { return receivedAmount_1 += Number(moneyIn.totalAmount); });
            balancedAmount = ((_45 = args.sale) === null || _45 === void 0 ? void 0 : _45.totalAmount) - receivedAmount_1;
            if (((_46 = args.profile) === null || _46 === void 0 ? void 0 : _46.pSetAlwaysQrPrintStatus) || (Number(balancedAmount) || 0) > 0) {
                if (balancedAmount == 0) {
                    balancedAmount = (_47 = args.sale) === null || _47 === void 0 ? void 0 : _47.totalAmount;
                }
                billPrint.qrPayLink = "upi://pay?pa=".concat((_48 = args.profile) === null || _48 === void 0 ? void 0 : _48.bankUPI, "&pn=").concat((_49 = args.profile) === null || _49 === void 0 ? void 0 : _49.legalName, "&am=").concat(balancedAmount, "&cu=INR&tn=").concat(args.sale.billNo);
            }
        }
        return billPrint;
    };
    HtmlTemplates.generateSaleBillObject = function (args) {
        var _this = this;
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36, _37, _38, _39, _40, _41, _42, _43, _44, _45, _46, _47, _48, _49;
        var billPrint = new SaleBillPrint_model_1.SaleBillPrint();
        billPrint.user = args.user;
        billPrint.profile = {
            addressProvience: (_a = args.profile) === null || _a === void 0 ? void 0 : _a.addressProvience,
            legalName: (_b = args.profile) === null || _b === void 0 ? void 0 : _b.legalName,
            addressLine1: (((_c = args.profile) === null || _c === void 0 ? void 0 : _c.addressLine1) || '') + (((_d = args.profile) === null || _d === void 0 ? void 0 : _d.addressLine2) || ''),
            addressPostalCode: (_e = args.profile) === null || _e === void 0 ? void 0 : _e.addressPostalCode,
            contactPersonPhone: (_f = args.profile) === null || _f === void 0 ? void 0 : _f.contactPersonPhone,
            contactPersonEmail: (_g = args.profile) === null || _g === void 0 ? void 0 : _g.contactPersonEmail,
            gstin: (_h = args.profile) === null || _h === void 0 ? void 0 : _h.gstin,
            fssaiNumber: (_j = args.profile) === null || _j === void 0 ? void 0 : _j.fssaiNumber,
            licenseNumber: (_k = args.profile) === null || _k === void 0 ? void 0 : _k.licenseNumber,
            bankAccountNo: (_l = args.profile) === null || _l === void 0 ? void 0 : _l.bankAccountNo,
            bankName: (_m = args.profile) === null || _m === void 0 ? void 0 : _m.bankName,
            bankIFSC: (_o = args.profile) === null || _o === void 0 ? void 0 : _o.bankIFSC,
            bankAccountType: (_p = args.profile) === null || _p === void 0 ? void 0 : _p.bankAccountType,
            bankUPI: (_q = args.profile) === null || _q === void 0 ? void 0 : _q.bankUPI,
            additionalInputFieldTitle: (_r = args.profile) === null || _r === void 0 ? void 0 : _r.additionalInputFieldTitle,
            additionalInputFieldValue: (_s = args.profile) === null || _s === void 0 ? void 0 : _s.additionalInputFieldValue,
            additionalDateFieldValue: (_t = args.profile) === null || _t === void 0 ? void 0 : _t.additionalDateFieldValue,
            additionalDateFieldTitle: (_u = args.profile) === null || _u === void 0 ? void 0 : _u.additionalDateFieldTitle,
            signatureLink: ((_v = args.signature) === null || _v === void 0 ? void 0 : _v.imageBase64) || '',
            logoLink: ((_w = args.profile) === null || _w === void 0 ? void 0 : _w.pSetLogoPrintStatus) ? ((_x = args.logo) === null || _x === void 0 ? void 0 : _x.imageBase64) || '' : '',
        };
        billPrint.bill = {
            partyData: {
                ledgerCredit: (_y = args.party) === null || _y === void 0 ? void 0 : _y.credit,
                profileData: {
                    legalName: (_z = args.party) === null || _z === void 0 ? void 0 : _z.businessName,
                    gstNumber: (_0 = args.party) === null || _0 === void 0 ? void 0 : _0.gstin,
                    contactPersonName: (_1 = args.party) === null || _1 === void 0 ? void 0 : _1.name,
                    contactPersonPhone: (_2 = args.party) === null || _2 === void 0 ? void 0 : _2.phone,
                    contactPersonEmail: (_3 = args.party) === null || _3 === void 0 ? void 0 : _3.email,
                    addressLine1: (_4 = args.party) === null || _4 === void 0 ? void 0 : _4.billingAddress,
                    addressState: (_5 = args.party) === null || _5 === void 0 ? void 0 : _5.billingProvience,
                    addressPincode: (_6 = args.party) === null || _6 === void 0 ? void 0 : _6.billingPostalCode,
                    addressOneLine1: (_7 = args.party) === null || _7 === void 0 ? void 0 : _7.deliveryAddress,
                    addressOneState: (_8 = args.party) === null || _8 === void 0 ? void 0 : _8.deliveryProvience,
                    addressOnePincode: (_9 = args.party) === null || _9 === void 0 ? void 0 : _9.deliveryPostalCode,
                    udf1T: '',
                    udf1V: '',
                    udf2T: '',
                    udf2V: '',
                    udf3T: '',
                    udf3V: '',
                    additionalInputFieldTitle: '',
                    additionalInputFieldValue: '',
                    additionalDateFieldValue: '',
                    additionalDateFieldTitle: '',
                },
            },
            moneyIn: {
                billNo: (_10 = args.moneyIn) === null || _10 === void 0 ? void 0 : _10.billNo,
                dateStamp: (_11 = args.moneyIn) === null || _11 === void 0 ? void 0 : _11.billDateStamp,
                amount: ((_12 = args.sale) === null || _12 === void 0 ? void 0 : _12.amountReceived) || ((_13 = args.moneyIn) === null || _13 === void 0 ? void 0 : _13.totalAmount),
                txnMode: (_14 = args.moneyIn) === null || _14 === void 0 ? void 0 : _14.paymentMode,
                txnRef: (_15 = args.moneyIn) === null || _15 === void 0 ? void 0 : _15.paymentId,
            },
            isPartyDeleted: args.isPartyDeleted,
            billItems: (_16 = args.sale.billItems) === null || _16 === void 0 ? void 0 : _16.map(function (billItem) {
                var _a, _b, _c, _d, _e, _f, _g;
                return {
                    itemDes: _this.newLineToBr((billItem === null || billItem === void 0 ? void 0 : billItem.billNote) || ''),
                    itemSrl: '',
                    hsn: ((_a = billItem.item) === null || _a === void 0 ? void 0 : _a.hsn) || '',
                    price: _this.capFractionsToTwo(billItem.price),
                    basePrice: _this.capFractionsToTwo(billItem.basePrice),
                    incTax: billItem.incTax,
                    taxPercentage: _this.capFractionsToTwo(billItem.taxPercentage),
                    cessPercentage: _this.capFractionsToTwo(billItem.cessPercentage),
                    unit: billItem.unit,
                    quantity: billItem.quantity,
                    discountPercentage: _this.capFractionsToTwo(billItem.discountPercentage),
                    discount: _this.capFractionsToTwo(billItem.discount),
                    unitGstAmount: _this.capFractionsToTwo(billItem.unitGstAmount),
                    unitCessAmount: _this.capFractionsToTwo(billItem.unitCessAmount),
                    unitDiscountAmount: _this.capFractionsToTwo(billItem.unitDiscountAmount),
                    unitTaxAmount: _this.capFractionsToTwo(billItem.unitTaxAmount),
                    itemTotalGstAmount: _this.capFractionsToTwo(billItem.itemTotalGstAmount),
                    itemTotalCessAmount: _this.capFractionsToTwo(billItem.itemTotalCessAmount),
                    itemTotalTaxAmount: _this.capFractionsToTwo(billItem.itemTotalTaxAmount),
                    effectiveTaxPercentage: _this.capFractionsToTwo(billItem.effectiveTaxPercentage),
                    billCashDiscountPercentage: _this.capFractionsToTwo(billItem.billCashDiscountPercentage),
                    total: _this.capFractionsToTwo(billItem.wcdTotal),
                    createdStamp: (_b = billItem.item) === null || _b === void 0 ? void 0 : _b.createdStamp,
                    item: {
                        itemName: (_c = billItem.item) === null || _c === void 0 ? void 0 : _c.itemName,
                        hsn: (_d = billItem.item) === null || _d === void 0 ? void 0 : _d.hsn,
                        mrp: (_e = billItem.item) === null || _e === void 0 ? void 0 : _e.mrp,
                        expiryDate: (_f = billItem.item) === null || _f === void 0 ? void 0 : _f.expiryDate,
                    },
                    additionalInputFieldValue: '',
                    additionalDateFieldValue: 0,
                    additionalInputFieldTitle: '',
                    additionalDateFieldTitle: '',
                    mrp: ((_g = billItem.item) === null || _g === void 0 ? void 0 : _g.mrp) || null,
                    variant: '',
                };
            }),
            roundOffValue: this.capFractionsToTwo(args.sale.roundOffValue) || 0,
            billNo: args.sale.billNo,
            deliveryState: args.sale.party.deliveryProvience,
            billDateStamp: args.sale.billDateStamp,
            dueDateStamp: args.sale.dueDateStamp,
            purOrderNo: (_17 = args.sale.transportDetail) === null || _17 === void 0 ? void 0 : _17.purOrderNo,
            eWayBillNo: (_18 = args.sale.transportDetail) === null || _18 === void 0 ? void 0 : _18.eWayBillNo,
            additionalAmount: this.capFractionsToTwo((_19 = args.sale.transportDetail) === null || _19 === void 0 ? void 0 : _19.additionalAmount),
            discountAmount: this.capFractionsToTwo(args.sale.discountAmount),
            gstAmount: this.capFractionsToTwo(args.sale.gstAmount),
            cessAmount: this.capFractionsToTwo(args.sale.cessAmount),
            totalAmount: this.capFractionsToTwo(args.sale.totalAmount),
            createdStamp: args.sale.createdStamp,
            transporterName: (_20 = args.sale.transportDetail) === null || _20 === void 0 ? void 0 : _20.transporterName,
            vehicleNumber: (_21 = args.sale.transportDetail) === null || _21 === void 0 ? void 0 : _21.vehicleNumber,
            deliveryDate: (_22 = args.sale.transportDetail) === null || _22 === void 0 ? void 0 : _22.deliveryDate,
            color: '#3498db',
            eWayBillDate: (_23 = args.sale.transportDetail) === null || _23 === void 0 ? void 0 : _23.eWayBillDate,
            note: args.sale.note,
            billTitle: '',
            tnc: ((_24 = args.profile) === null || _24 === void 0 ? void 0 : _24.pSetTermsAndConditions) || null,
            cashDiscount: this.capFractionsToTwo(args.sale.cashDiscount),
            amountReceived: this.capFractionsToTwo(args.sale.amountReceived),
            cashDiscountPercentage: this.capFractionsToTwo(args.sale.cashDiscountPercentage),
            challanNo: ((_25 = args.sale.transportDetail) === null || _25 === void 0 ? void 0 : _25.challanNo) || null,
            totalSaving: this.capFractionsToTwo(args.sale.totalSaving),
            subTotalAmount: this.capFractionsToTwo(args.sale.subTotalAmount),
            partyPreviousBalance: this.capFractionsToTwo(args.sale.partyPreviousBalance) || null,
            billCompleteStamp: args.sale.billCompleteStamp,
            serviceChargePercentage: this.capFractionsToTwo(args.sale.serviceChargePercentage),
            serviceChargeAmount: this.capFractionsToTwo(args.sale.serviceChargeAmount),
            deliveryProvience: (_26 = args === null || args === void 0 ? void 0 : args.sale) === null || _26 === void 0 ? void 0 : _26.deliveryProvience,
        };
        if ((_27 = args.secondaryParty) === null || _27 === void 0 ? void 0 : _27._localUUID) {
            billPrint.bill.secondaryPartyData = {
                ledgerCredit: (_28 = args.secondaryParty) === null || _28 === void 0 ? void 0 : _28.credit,
                profileData: {
                    legalName: (_29 = args.secondaryParty) === null || _29 === void 0 ? void 0 : _29.businessName,
                    gstNumber: (_30 = args.secondaryParty) === null || _30 === void 0 ? void 0 : _30.gstin,
                    contactPersonName: (_31 = args.secondaryParty) === null || _31 === void 0 ? void 0 : _31.name,
                    contactPersonPhone: (_32 = args.secondaryParty) === null || _32 === void 0 ? void 0 : _32.phone,
                    contactPersonEmail: (_33 = args.secondaryParty) === null || _33 === void 0 ? void 0 : _33.email,
                    addressLine1: (_34 = args.secondaryParty) === null || _34 === void 0 ? void 0 : _34.billingAddress,
                    addressState: (_35 = args.secondaryParty) === null || _35 === void 0 ? void 0 : _35.billingProvience,
                    addressPincode: (_36 = args.secondaryParty) === null || _36 === void 0 ? void 0 : _36.billingPostalCode,
                    addressOneLine1: (_37 = args.secondaryParty) === null || _37 === void 0 ? void 0 : _37.deliveryAddress,
                    addressOneState: (_38 = args.secondaryParty) === null || _38 === void 0 ? void 0 : _38.deliveryProvience,
                    addressOnePincode: (_39 = args.secondaryParty) === null || _39 === void 0 ? void 0 : _39.deliveryPostalCode,
                    udf1T: '',
                    udf1V: '',
                    udf2T: '',
                    udf2V: '',
                    udf3T: '',
                    udf3V: '',
                    additionalInputFieldTitle: '',
                    additionalInputFieldValue: '',
                    additionalDateFieldValue: '',
                    additionalDateFieldTitle: '',
                },
            };
        }
        // set settings - TODO All
        billPrint.settings = {
            showSignatureNotRequiredText: false,
            showSignatureInTemplates: true,
            makeItemNameBold: false,
            showPreviousAmount: false,
            showTaxAndDiscAmt: (_40 = args.profile) === null || _40 === void 0 ? void 0 : _40.pSetTaxMrpHsnPrintStatus,
            billSavingsAmount: true,
        };
        // set extra - TODO All
        billPrint.extra = {
            thermalTncTemplate: ((_41 = args.profile) === null || _41 === void 0 ? void 0 : _41.pSetTermsAndConditions) || null,
        };
        billPrint.qrPayLink = '';
        billPrint.addressTo = '';
        billPrint.pageWidth = null;
        if (((_42 = args.profile) === null || _42 === void 0 ? void 0 : _42.bankUPI) && ((_43 = args.profile) === null || _43 === void 0 ? void 0 : _43.pSetQrPrintStatus)) {
            var receivedAmount_2 = 0;
            var balancedAmount = 0;
            (_44 = args.sale) === null || _44 === void 0 ? void 0 : _44.moneyIns.forEach(function (moneyIn) { return receivedAmount_2 += Number(moneyIn.totalAmount); });
            balancedAmount = ((_45 = args.sale) === null || _45 === void 0 ? void 0 : _45.totalAmount) - receivedAmount_2;
            if (((_46 = args.profile) === null || _46 === void 0 ? void 0 : _46.pSetAlwaysQrPrintStatus) || (Number(balancedAmount) || 0) > 0) {
                if (balancedAmount == 0) {
                    balancedAmount = (_47 = args.sale) === null || _47 === void 0 ? void 0 : _47.totalAmount;
                }
                billPrint.qrPayLink = "upi://pay?pa=".concat((_48 = args.profile) === null || _48 === void 0 ? void 0 : _48.bankUPI, "&pn=").concat((_49 = args.profile) === null || _49 === void 0 ? void 0 : _49.legalName, "&am=").concat(balancedAmount, "&cu=INR&tn=").concat(args.sale.billNo);
            }
        }
        return billPrint;
    };
    HtmlTemplates.generatePurchaseBillObject = function (args) {
        var _this = this;
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36;
        var billPrint = new PurchaseBillPrint_model_1.PurchaseBillPrint();
        billPrint.user = args.user;
        billPrint.profile = {
            addressProvience: (_a = args.profile) === null || _a === void 0 ? void 0 : _a.addressProvience,
            legalName: (_b = args.profile) === null || _b === void 0 ? void 0 : _b.legalName,
            addressLine1: (((_c = args.profile) === null || _c === void 0 ? void 0 : _c.addressLine1) || '') + (((_d = args.profile) === null || _d === void 0 ? void 0 : _d.addressLine2) || ''),
            addressPostalCode: (_e = args.profile) === null || _e === void 0 ? void 0 : _e.addressPostalCode,
            contactPersonPhone: (_f = args.profile) === null || _f === void 0 ? void 0 : _f.contactPersonPhone,
            contactPersonEmail: (_g = args.profile) === null || _g === void 0 ? void 0 : _g.contactPersonEmail,
            gstin: (_h = args.profile) === null || _h === void 0 ? void 0 : _h.gstin,
            fssaiNumber: (_j = args.profile) === null || _j === void 0 ? void 0 : _j.fssaiNumber,
            licenseNumber: (_k = args.profile) === null || _k === void 0 ? void 0 : _k.licenseNumber,
            bankAccountNo: (_l = args.profile) === null || _l === void 0 ? void 0 : _l.bankAccountNo,
            bankName: (_m = args.profile) === null || _m === void 0 ? void 0 : _m.bankName,
            bankIFSC: (_o = args.profile) === null || _o === void 0 ? void 0 : _o.bankIFSC,
            bankAccountType: (_p = args.profile) === null || _p === void 0 ? void 0 : _p.bankAccountType,
            bankUPI: (_q = args.profile) === null || _q === void 0 ? void 0 : _q.bankUPI,
            additionalInputFieldTitle: (_r = args.profile) === null || _r === void 0 ? void 0 : _r.additionalInputFieldTitle,
            additionalInputFieldValue: (_s = args.profile) === null || _s === void 0 ? void 0 : _s.additionalInputFieldValue,
            additionalDateFieldValue: (_t = args.profile) === null || _t === void 0 ? void 0 : _t.additionalDateFieldValue,
            additionalDateFieldTitle: (_u = args.profile) === null || _u === void 0 ? void 0 : _u.additionalDateFieldTitle,
            signatureLink: ((_v = args.signature) === null || _v === void 0 ? void 0 : _v.imageBase64) || '',
            logoLink: ((_w = args.profile) === null || _w === void 0 ? void 0 : _w.pSetLogoPrintStatus) ? ((_x = args.logo) === null || _x === void 0 ? void 0 : _x.imageBase64) || '' : '',
        };
        billPrint.bill = {
            partyData: {
                ledgerCredit: (_y = args.party) === null || _y === void 0 ? void 0 : _y.credit,
                profileData: {
                    legalName: (_z = args.party) === null || _z === void 0 ? void 0 : _z.businessName,
                    gstNumber: (_0 = args.party) === null || _0 === void 0 ? void 0 : _0.gstin,
                    contactPersonName: (_1 = args.party) === null || _1 === void 0 ? void 0 : _1.name,
                    contactPersonPhone: (_2 = args.party) === null || _2 === void 0 ? void 0 : _2.phone,
                    contactPersonEmail: (_3 = args.party) === null || _3 === void 0 ? void 0 : _3.email,
                    addressLine1: (_4 = args.party) === null || _4 === void 0 ? void 0 : _4.billingAddress,
                    addressState: (_5 = args.party) === null || _5 === void 0 ? void 0 : _5.billingProvience,
                    addressPincode: (_6 = args.party) === null || _6 === void 0 ? void 0 : _6.billingPostalCode,
                    addressOneLine1: (_7 = args.party) === null || _7 === void 0 ? void 0 : _7.deliveryAddress,
                    addressOneState: (_8 = args.party) === null || _8 === void 0 ? void 0 : _8.deliveryProvience,
                    addressOnePincode: (_9 = args.party) === null || _9 === void 0 ? void 0 : _9.deliveryPostalCode,
                    udf1T: '',
                    udf1V: '',
                    udf2T: '',
                    udf2V: '',
                    udf3T: '',
                    udf3V: '',
                    additionalInputFieldTitle: '',
                    additionalInputFieldValue: '',
                    additionalDateFieldValue: '',
                    additionalDateFieldTitle: '',
                },
            },
            moneyOut: {
                billNo: (_10 = args.moneyOut) === null || _10 === void 0 ? void 0 : _10.billNo,
                dateStamp: (_11 = args.moneyOut) === null || _11 === void 0 ? void 0 : _11.billDateStamp,
                amount: ((_12 = args.purchase) === null || _12 === void 0 ? void 0 : _12.amountPaid) || ((_13 = args.moneyOut) === null || _13 === void 0 ? void 0 : _13.totalAmount),
                txnMode: (_14 = args.moneyOut) === null || _14 === void 0 ? void 0 : _14.paymentMode,
                txnRef: (_15 = args.moneyOut) === null || _15 === void 0 ? void 0 : _15.paymentId,
            },
            billItems: (_16 = args.purchase.billItems) === null || _16 === void 0 ? void 0 : _16.map(function (billItem) {
                var _a, _b, _c, _d, _e, _f, _g;
                return {
                    itemDes: _this.newLineToBr((billItem === null || billItem === void 0 ? void 0 : billItem.billNote) || ''),
                    itemSrl: '',
                    hsn: ((_a = billItem.item) === null || _a === void 0 ? void 0 : _a.hsn) || '',
                    price: _this.capFractionsToTwo(billItem.price),
                    basePrice: _this.capFractionsToTwo(billItem.basePrice),
                    incTax: billItem.incTax,
                    taxPercentage: _this.capFractionsToTwo(billItem.taxPercentage),
                    cessPercentage: _this.capFractionsToTwo(billItem.cessPercentage),
                    unit: billItem.unit,
                    quantity: billItem.quantity,
                    discountPercentage: _this.capFractionsToTwo(billItem.discountPercentage),
                    discount: _this.capFractionsToTwo(billItem.discount),
                    unitGstAmount: _this.capFractionsToTwo(billItem.unitGstAmount),
                    unitCessAmount: _this.capFractionsToTwo(billItem.unitCessAmount),
                    unitDiscountAmount: _this.capFractionsToTwo(billItem.unitDiscountAmount),
                    unitTaxAmount: _this.capFractionsToTwo(billItem.unitTaxAmount),
                    itemTotalGstAmount: _this.capFractionsToTwo(billItem.itemTotalGstAmount),
                    itemTotalCessAmount: _this.capFractionsToTwo(billItem.itemTotalCessAmount),
                    itemTotalTaxAmount: _this.capFractionsToTwo(billItem.itemTotalTaxAmount),
                    effectiveTaxPercentage: _this.capFractionsToTwo(billItem.effectiveTaxPercentage),
                    billCashDiscountPercentage: _this.capFractionsToTwo(billItem.billCashDiscountPercentage),
                    total: _this.capFractionsToTwo(billItem.wcdTotal),
                    createdStamp: (_b = billItem.item) === null || _b === void 0 ? void 0 : _b.createdStamp,
                    item: {
                        itemName: (_c = billItem.item) === null || _c === void 0 ? void 0 : _c.itemName,
                        hsn: (_d = billItem.item) === null || _d === void 0 ? void 0 : _d.hsn,
                        mrp: (_e = billItem.item) === null || _e === void 0 ? void 0 : _e.mrp,
                        expiryDate: (_f = billItem.item) === null || _f === void 0 ? void 0 : _f.expiryDate,
                    },
                    additionalInputFieldValue: '',
                    additionalDateFieldValue: 0,
                    additionalInputFieldTitle: '',
                    additionalDateFieldTitle: '',
                    mrp: ((_g = billItem.item) === null || _g === void 0 ? void 0 : _g.mrp) || null,
                    variant: '',
                };
            }),
            roundOffValue: this.capFractionsToTwo(args.purchase.roundOffValue) || 0,
            billNo: args.purchase.billNo,
            deliveryState: args.purchase.party.deliveryProvience,
            billDateStamp: args.purchase.billDateStamp,
            dueDateStamp: args.purchase.dueDateStamp,
            purOrderNo: (_17 = args.purchase.transportDetail) === null || _17 === void 0 ? void 0 : _17.purOrderNo,
            eWayBillNo: (_18 = args.purchase.transportDetail) === null || _18 === void 0 ? void 0 : _18.eWayBillNo,
            additionalAmount: this.capFractionsToTwo((_19 = args.purchase.transportDetail) === null || _19 === void 0 ? void 0 : _19.additionalAmount),
            discountAmount: this.capFractionsToTwo(args.purchase.discountAmount),
            gstAmount: this.capFractionsToTwo(args.purchase.gstAmount),
            cessAmount: this.capFractionsToTwo(args.purchase.cessAmount),
            totalAmount: this.capFractionsToTwo(args.purchase.totalAmount),
            createdStamp: args.purchase.createdStamp,
            transporterName: (_20 = args.purchase.transportDetail) === null || _20 === void 0 ? void 0 : _20.transporterName,
            vehicleNumber: (_21 = args.purchase.transportDetail) === null || _21 === void 0 ? void 0 : _21.vehicleNumber,
            deliveryDate: (_22 = args.purchase.transportDetail) === null || _22 === void 0 ? void 0 : _22.deliveryDate,
            color: '#3498db',
            eWayBillDate: (_23 = args.purchase.transportDetail) === null || _23 === void 0 ? void 0 : _23.eWayBillDate,
            note: args.purchase.note,
            billTitle: '',
            tnc: ((_24 = args.profile) === null || _24 === void 0 ? void 0 : _24.pSetTermsAndConditions) || null,
            cashDiscount: this.capFractionsToTwo(args.purchase.cashDiscount),
            amountPaid: this.capFractionsToTwo(args.purchase.amountPaid),
            cashDiscountPercentage: this.capFractionsToTwo(args.purchase.cashDiscountPercentage),
            challanNo: ((_25 = args.purchase.transportDetail) === null || _25 === void 0 ? void 0 : _25.challanNo) || null,
            totalSaving: this.capFractionsToTwo(args.purchase.totalSaving),
            subTotalAmount: this.capFractionsToTwo(args.purchase.subTotalAmount),
            partyPreviousBalance: this.capFractionsToTwo(args.purchase.partyPreviousBalance) || null,
            billCompleteStamp: args.purchase.billCompleteStamp,
            serviceChargePercentage: this.capFractionsToTwo(args.purchase.serviceChargePercentage),
            serviceChargeAmount: this.capFractionsToTwo(args.purchase.serviceChargeAmount),
            deliveryProvience: (_26 = args === null || args === void 0 ? void 0 : args.purchase) === null || _26 === void 0 ? void 0 : _26.deliveryProvience,
        };
        // set settings - TODO All
        billPrint.settings = {
            showSignatureNotRequiredText: false,
            showSignatureInTemplates: true,
            makeItemNameBold: false,
            showPreviousAmount: false,
            showTaxAndDiscAmt: (_27 = args.profile) === null || _27 === void 0 ? void 0 : _27.pSetTaxMrpHsnPrintStatus,
            billSavingsAmount: true,
        };
        // set extra - TODO All
        billPrint.extra = {
            thermalTncTemplate: ((_28 = args.profile) === null || _28 === void 0 ? void 0 : _28.pSetTermsAndConditions) || null,
        };
        billPrint.qrPayLink = '';
        billPrint.addressTo = '';
        billPrint.pageWidth = null;
        if (((_29 = args.profile) === null || _29 === void 0 ? void 0 : _29.bankUPI) && ((_30 = args.profile) === null || _30 === void 0 ? void 0 : _30.pSetQrPrintStatus)) {
            var receivedAmount_3 = 0;
            var balancedAmount = 0;
            (_31 = args.purchase) === null || _31 === void 0 ? void 0 : _31.moneyOuts.forEach(function (moneyOut) { return receivedAmount_3 += Number(moneyOut.totalAmount); });
            balancedAmount = ((_32 = args.purchase) === null || _32 === void 0 ? void 0 : _32.totalAmount) - receivedAmount_3;
            if (((_33 = args.profile) === null || _33 === void 0 ? void 0 : _33.pSetAlwaysQrPrintStatus) || (Number(balancedAmount) || 0) > 0) {
                if (balancedAmount == 0) {
                    balancedAmount = (_34 = args.purchase) === null || _34 === void 0 ? void 0 : _34.totalAmount;
                }
                billPrint.qrPayLink = "upi://pay?pa=".concat((_35 = args.profile) === null || _35 === void 0 ? void 0 : _35.bankUPI, "&pn=").concat((_36 = args.profile) === null || _36 === void 0 ? void 0 : _36.legalName, "&am=").concat(balancedAmount, "&cu=INR&tn=").concat(args.purchase.billNo);
            }
        }
        return billPrint;
    };
    HtmlTemplates.generateMoneyInBillObject = function (args) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13;
        var billPrint = new MoneyInBillPrint_model_1.MoneyInBillPrint();
        billPrint.user = args.user;
        billPrint.profile = {
            addressProvience: (_a = args.profile) === null || _a === void 0 ? void 0 : _a.addressProvience,
            legalName: (_b = args.profile) === null || _b === void 0 ? void 0 : _b.legalName,
            addressLine1: (((_c = args.profile) === null || _c === void 0 ? void 0 : _c.addressLine1) || '') + (((_d = args.profile) === null || _d === void 0 ? void 0 : _d.addressLine2) || ''),
            addressPostalCode: (_e = args.profile) === null || _e === void 0 ? void 0 : _e.addressPostalCode,
            contactPersonPhone: (_f = args.profile) === null || _f === void 0 ? void 0 : _f.contactPersonPhone,
            contactPersonEmail: (_g = args.profile) === null || _g === void 0 ? void 0 : _g.contactPersonEmail,
            gstin: (_h = args.profile) === null || _h === void 0 ? void 0 : _h.gstin,
            fssaiNumber: (_j = args.profile) === null || _j === void 0 ? void 0 : _j.fssaiNumber,
            licenseNumber: (_k = args.profile) === null || _k === void 0 ? void 0 : _k.licenseNumber,
            bankAccountNo: (_l = args.profile) === null || _l === void 0 ? void 0 : _l.bankAccountNo,
            bankName: (_m = args.profile) === null || _m === void 0 ? void 0 : _m.bankName,
            bankIFSC: (_o = args.profile) === null || _o === void 0 ? void 0 : _o.bankIFSC,
            bankAccountType: (_p = args.profile) === null || _p === void 0 ? void 0 : _p.bankAccountType,
            bankUPI: (_q = args.profile) === null || _q === void 0 ? void 0 : _q.bankUPI,
            additionalInputFieldTitle: (_r = args.profile) === null || _r === void 0 ? void 0 : _r.additionalInputFieldTitle,
            additionalInputFieldValue: (_s = args.profile) === null || _s === void 0 ? void 0 : _s.additionalInputFieldValue,
            additionalDateFieldValue: (_t = args.profile) === null || _t === void 0 ? void 0 : _t.additionalDateFieldValue,
            additionalDateFieldTitle: (_u = args.profile) === null || _u === void 0 ? void 0 : _u.additionalDateFieldTitle,
            signatureLink: ((_v = args.signature) === null || _v === void 0 ? void 0 : _v.imageBase64) || '',
            logoLink: ((_w = args.profile) === null || _w === void 0 ? void 0 : _w.pSetLogoPrintStatus) ? ((_x = args.logo) === null || _x === void 0 ? void 0 : _x.imageBase64) || '' : '',
        };
        billPrint.bill = {
            partyData: {
                ledgerCredit: (_y = args.party) === null || _y === void 0 ? void 0 : _y.credit,
                profileData: {
                    legalName: (_z = args.party) === null || _z === void 0 ? void 0 : _z.businessName,
                    gstNumber: (_0 = args.party) === null || _0 === void 0 ? void 0 : _0.gstin,
                    contactPersonName: (_1 = args.party) === null || _1 === void 0 ? void 0 : _1.name,
                    contactPersonPhone: (_2 = args.party) === null || _2 === void 0 ? void 0 : _2.phone,
                    contactPersonEmail: (_3 = args.party) === null || _3 === void 0 ? void 0 : _3.email,
                    addressLine1: (_4 = args.party) === null || _4 === void 0 ? void 0 : _4.billingAddress,
                    addressState: (_5 = args.party) === null || _5 === void 0 ? void 0 : _5.billingProvience,
                    addressPincode: (_6 = args.party) === null || _6 === void 0 ? void 0 : _6.billingPostalCode,
                    addressOneLine1: (_7 = args.party) === null || _7 === void 0 ? void 0 : _7.deliveryAddress,
                    addressOneState: (_8 = args.party) === null || _8 === void 0 ? void 0 : _8.deliveryProvience,
                    addressOnePincode: (_9 = args.party) === null || _9 === void 0 ? void 0 : _9.deliveryPostalCode,
                    udf1T: '',
                    udf1V: '',
                    udf2T: '',
                    udf2V: '',
                    udf3T: '',
                    udf3V: '',
                    additionalInputFieldTitle: '',
                    additionalInputFieldValue: '',
                    additionalDateFieldValue: '',
                    additionalDateFieldTitle: '',
                },
            },
            txnMode: (_10 = args.moneyIn) === null || _10 === void 0 ? void 0 : _10.paymentMode,
            txnRef: (_11 = args.moneyIn) === null || _11 === void 0 ? void 0 : _11.paymentId,
            isPartyDeleted: args.isPartyDeleted,
            billNo: args.moneyIn.billNo,
            billDateStamp: args.moneyIn.billDateStamp,
            totalAmount: this.capFractionsToTwo(args.moneyIn.totalAmount),
            createdStamp: args.moneyIn.createdStamp,
            color: '#3498db',
            billTitle: '',
            tnc: ((_12 = args.profile) === null || _12 === void 0 ? void 0 : _12.pSetTermsAndConditions) || null,
        };
        // set settings - TODO All
        billPrint.settings = {
            showSignatureNotRequiredText: false,
            showSignatureInTemplates: true,
        };
        // set extra - TODO All
        billPrint.extra = {
            thermalTncTemplate: ((_13 = args.profile) === null || _13 === void 0 ? void 0 : _13.pSetTermsAndConditions) || null,
        };
        billPrint.qrPayLink = '';
        billPrint.addressTo = '';
        billPrint.pageWidth = null;
        return billPrint;
    };
    HtmlTemplates.generateMoneyOutBillObject = function (args) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13;
        var billPrint = new MoneyOutBillPrint_model_1.MoneyOutBillPrint();
        billPrint.user = args.user;
        billPrint.profile = {
            addressProvience: (_a = args.profile) === null || _a === void 0 ? void 0 : _a.addressProvience,
            legalName: (_b = args.profile) === null || _b === void 0 ? void 0 : _b.legalName,
            addressLine1: (((_c = args.profile) === null || _c === void 0 ? void 0 : _c.addressLine1) || '') + (((_d = args.profile) === null || _d === void 0 ? void 0 : _d.addressLine2) || ''),
            addressPostalCode: (_e = args.profile) === null || _e === void 0 ? void 0 : _e.addressPostalCode,
            contactPersonPhone: (_f = args.profile) === null || _f === void 0 ? void 0 : _f.contactPersonPhone,
            contactPersonEmail: (_g = args.profile) === null || _g === void 0 ? void 0 : _g.contactPersonEmail,
            gstin: (_h = args.profile) === null || _h === void 0 ? void 0 : _h.gstin,
            fssaiNumber: (_j = args.profile) === null || _j === void 0 ? void 0 : _j.fssaiNumber,
            licenseNumber: (_k = args.profile) === null || _k === void 0 ? void 0 : _k.licenseNumber,
            bankAccountNo: (_l = args.profile) === null || _l === void 0 ? void 0 : _l.bankAccountNo,
            bankName: (_m = args.profile) === null || _m === void 0 ? void 0 : _m.bankName,
            bankIFSC: (_o = args.profile) === null || _o === void 0 ? void 0 : _o.bankIFSC,
            bankAccountType: (_p = args.profile) === null || _p === void 0 ? void 0 : _p.bankAccountType,
            bankUPI: (_q = args.profile) === null || _q === void 0 ? void 0 : _q.bankUPI,
            additionalInputFieldTitle: (_r = args.profile) === null || _r === void 0 ? void 0 : _r.additionalInputFieldTitle,
            additionalInputFieldValue: (_s = args.profile) === null || _s === void 0 ? void 0 : _s.additionalInputFieldValue,
            additionalDateFieldValue: (_t = args.profile) === null || _t === void 0 ? void 0 : _t.additionalDateFieldValue,
            additionalDateFieldTitle: (_u = args.profile) === null || _u === void 0 ? void 0 : _u.additionalDateFieldTitle,
            signatureLink: ((_v = args.signature) === null || _v === void 0 ? void 0 : _v.imageBase64) || '',
            logoLink: ((_w = args.profile) === null || _w === void 0 ? void 0 : _w.pSetLogoPrintStatus) ? ((_x = args.logo) === null || _x === void 0 ? void 0 : _x.imageBase64) || '' : '',
        };
        billPrint.bill = {
            partyData: {
                ledgerCredit: (_y = args.party) === null || _y === void 0 ? void 0 : _y.credit,
                profileData: {
                    legalName: (_z = args.party) === null || _z === void 0 ? void 0 : _z.businessName,
                    gstNumber: (_0 = args.party) === null || _0 === void 0 ? void 0 : _0.gstin,
                    contactPersonName: (_1 = args.party) === null || _1 === void 0 ? void 0 : _1.name,
                    contactPersonPhone: (_2 = args.party) === null || _2 === void 0 ? void 0 : _2.phone,
                    contactPersonEmail: (_3 = args.party) === null || _3 === void 0 ? void 0 : _3.email,
                    addressLine1: (_4 = args.party) === null || _4 === void 0 ? void 0 : _4.billingAddress,
                    addressState: (_5 = args.party) === null || _5 === void 0 ? void 0 : _5.billingProvience,
                    addressPincode: (_6 = args.party) === null || _6 === void 0 ? void 0 : _6.billingPostalCode,
                    addressOneLine1: (_7 = args.party) === null || _7 === void 0 ? void 0 : _7.deliveryAddress,
                    addressOneState: (_8 = args.party) === null || _8 === void 0 ? void 0 : _8.deliveryProvience,
                    addressOnePincode: (_9 = args.party) === null || _9 === void 0 ? void 0 : _9.deliveryPostalCode,
                    udf1T: '',
                    udf1V: '',
                    udf2T: '',
                    udf2V: '',
                    udf3T: '',
                    udf3V: '',
                    additionalInputFieldTitle: '',
                    additionalInputFieldValue: '',
                    additionalDateFieldValue: '',
                    additionalDateFieldTitle: '',
                },
            },
            txnMode: (_10 = args.moneyOut) === null || _10 === void 0 ? void 0 : _10.paymentMode,
            txnRef: (_11 = args.moneyOut) === null || _11 === void 0 ? void 0 : _11.paymentId,
            isPartyDeleted: args.isPartyDeleted,
            billNo: args.moneyOut.billNo,
            billDateStamp: args.moneyOut.billDateStamp,
            totalAmount: this.capFractionsToTwo(args.moneyOut.totalAmount),
            createdStamp: args.moneyOut.createdStamp,
            color: '#3498db',
            billTitle: '',
            tnc: ((_12 = args.profile) === null || _12 === void 0 ? void 0 : _12.pSetTermsAndConditions) || null,
        };
        billPrint.settings = {
            showSignatureNotRequiredText: false,
            showSignatureInTemplates: true,
        };
        // set extra - TODO All
        billPrint.extra = {
            thermalTncTemplate: ((_13 = args.profile) === null || _13 === void 0 ? void 0 : _13.pSetTermsAndConditions) || null,
        };
        billPrint.qrPayLink = '';
        billPrint.addressTo = '';
        billPrint.pageWidth = null;
        return billPrint;
    };
    HtmlTemplates.generateExpenseBillObject = function (args) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25;
        var billPrint = new ExpenseBillPrint_model_1.ExpenseBillPrint();
        billPrint.user = args.user;
        //Category
        billPrint.category = args.expense.category;
        billPrint.profile = {
            addressProvience: (_a = args.profile) === null || _a === void 0 ? void 0 : _a.addressProvience,
            legalName: (_b = args.profile) === null || _b === void 0 ? void 0 : _b.legalName,
            addressLine1: (((_c = args.profile) === null || _c === void 0 ? void 0 : _c.addressLine1) || '') + (((_d = args.profile) === null || _d === void 0 ? void 0 : _d.addressLine2) || ''),
            addressPostalCode: (_e = args.profile) === null || _e === void 0 ? void 0 : _e.addressPostalCode,
            contactPersonPhone: (_f = args.profile) === null || _f === void 0 ? void 0 : _f.contactPersonPhone,
            contactPersonEmail: (_g = args.profile) === null || _g === void 0 ? void 0 : _g.contactPersonEmail,
            gstin: (_h = args.profile) === null || _h === void 0 ? void 0 : _h.gstin,
            fssaiNumber: (_j = args.profile) === null || _j === void 0 ? void 0 : _j.fssaiNumber,
            licenseNumber: (_k = args.profile) === null || _k === void 0 ? void 0 : _k.licenseNumber,
            bankAccountNo: (_l = args.profile) === null || _l === void 0 ? void 0 : _l.bankAccountNo,
            bankName: (_m = args.profile) === null || _m === void 0 ? void 0 : _m.bankName,
            bankIFSC: (_o = args.profile) === null || _o === void 0 ? void 0 : _o.bankIFSC,
            bankAccountType: (_p = args.profile) === null || _p === void 0 ? void 0 : _p.bankAccountType,
            bankUPI: (_q = args.profile) === null || _q === void 0 ? void 0 : _q.bankUPI,
            additionalInputFieldTitle: (_r = args.profile) === null || _r === void 0 ? void 0 : _r.additionalInputFieldTitle,
            additionalInputFieldValue: (_s = args.profile) === null || _s === void 0 ? void 0 : _s.additionalInputFieldValue,
            additionalDateFieldValue: (_t = args.profile) === null || _t === void 0 ? void 0 : _t.additionalDateFieldValue,
            additionalDateFieldTitle: (_u = args.profile) === null || _u === void 0 ? void 0 : _u.additionalDateFieldTitle,
            signatureLink: ((_v = args.signature) === null || _v === void 0 ? void 0 : _v.imageBase64) || '',
            logoLink: ((_w = args.profile) === null || _w === void 0 ? void 0 : _w.pSetLogoPrintStatus) ? ((_x = args.logo) === null || _x === void 0 ? void 0 : _x.imageBase64) || '' : '',
        };
        billPrint.bill = {
            partyData: {
                ledgerCredit: (_y = args.party) === null || _y === void 0 ? void 0 : _y.credit,
                profileData: {
                    legalName: (_z = args.party) === null || _z === void 0 ? void 0 : _z.businessName,
                    gstNumber: (_0 = args.party) === null || _0 === void 0 ? void 0 : _0.gstin,
                    contactPersonName: (_1 = args.party) === null || _1 === void 0 ? void 0 : _1.name,
                    contactPersonPhone: (_2 = args.party) === null || _2 === void 0 ? void 0 : _2.phone,
                    contactPersonEmail: (_3 = args.party) === null || _3 === void 0 ? void 0 : _3.email,
                    addressLine1: (_4 = args.party) === null || _4 === void 0 ? void 0 : _4.billingAddress,
                    addressState: (_5 = args.party) === null || _5 === void 0 ? void 0 : _5.billingProvience,
                    addressPincode: (_6 = args.party) === null || _6 === void 0 ? void 0 : _6.billingPostalCode,
                    addressOneLine1: (_7 = args.party) === null || _7 === void 0 ? void 0 : _7.deliveryAddress,
                    addressOneState: (_8 = args.party) === null || _8 === void 0 ? void 0 : _8.deliveryProvience,
                    addressOnePincode: (_9 = args.party) === null || _9 === void 0 ? void 0 : _9.deliveryPostalCode,
                    udf1T: '',
                    udf1V: '',
                    udf2T: '',
                    udf2V: '',
                    udf3T: '',
                    udf3V: '',
                    additionalInputFieldTitle: '',
                    additionalInputFieldValue: '',
                    additionalDateFieldValue: '',
                    additionalDateFieldTitle: '',
                },
            },
            isPartyDeleted: args.isPartyDeleted,
            moneyOut: {
                billNo: (_10 = args.moneyOut) === null || _10 === void 0 ? void 0 : _10.billNo,
                dateStamp: (_11 = args.moneyOut) === null || _11 === void 0 ? void 0 : _11.billDateStamp,
                amount: ((_12 = args.expense) === null || _12 === void 0 ? void 0 : _12.amountPaid) || ((_13 = args.moneyOut) === null || _13 === void 0 ? void 0 : _13.totalAmount),
                txnMode: (_14 = args.moneyOut) === null || _14 === void 0 ? void 0 : _14.paymentMode,
                txnRef: (_15 = args.moneyOut) === null || _15 === void 0 ? void 0 : _15.paymentId,
            },
            billNo: args.expense.billNo,
            billDateStamp: args.expense.billDateStamp,
            dueDateStamp: args.expense.dueDateStamp,
            totalAmount: this.capFractionsToTwo(args.expense.totalAmount),
            createdStamp: args.expense.createdStamp,
            color: '#3498db',
            billTitle: '',
            tnc: ((_16 = args.profile) === null || _16 === void 0 ? void 0 : _16.pSetTermsAndConditions) || null,
            partyPreviousBalance: this.capFractionsToTwo(args.expense.partyPreviousBalance) || null,
            billCompleteStamp: args.expense.billCompleteStamp,
        };
        // set settings - TODO All
        billPrint.settings = {
            showSignatureNotRequiredText: false,
            showSignatureInTemplates: true,
        };
        // set extra - TODO All
        billPrint.extra = {
            thermalTncTemplate: ((_17 = args.profile) === null || _17 === void 0 ? void 0 : _17.pSetTermsAndConditions) || null,
        };
        billPrint.qrPayLink = '';
        billPrint.addressTo = '';
        billPrint.pageWidth = null;
        if (((_18 = args.profile) === null || _18 === void 0 ? void 0 : _18.bankUPI) && ((_19 = args.profile) === null || _19 === void 0 ? void 0 : _19.pSetQrPrintStatus)) {
            var receivedAmount_4 = 0;
            var balancedAmount = 0;
            (_20 = args.expense) === null || _20 === void 0 ? void 0 : _20.moneyOuts.forEach(function (moneyOut) { return receivedAmount_4 += Number(moneyOut.totalAmount); });
            balancedAmount = ((_21 = args.expense) === null || _21 === void 0 ? void 0 : _21.totalAmount) - receivedAmount_4;
            if (((_22 = args.profile) === null || _22 === void 0 ? void 0 : _22.pSetAlwaysQrPrintStatus) || (Number(balancedAmount) || 0) > 0) {
                if (balancedAmount == 0) {
                    balancedAmount = (_23 = args.expense) === null || _23 === void 0 ? void 0 : _23.totalAmount;
                }
                billPrint.qrPayLink = "upi://pay?pa=".concat((_24 = args.profile) === null || _24 === void 0 ? void 0 : _24.bankUPI, "&pn=").concat((_25 = args.profile) === null || _25 === void 0 ? void 0 : _25.legalName, "&am=").concat(balancedAmount, "&cu=INR&tn=").concat(args.expense.billNo);
            }
        }
        return billPrint;
    };
    HtmlTemplates.generateEstimateBillObject = function (args) {
        var _this = this;
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36, _37, _38, _39, _40, _41, _42;
        var billPrint = new EstimateBillPrint_model_1.EstimateBillPrint();
        billPrint.user = args.user;
        billPrint.profile = {
            addressProvience: (_a = args.profile) === null || _a === void 0 ? void 0 : _a.addressProvience,
            legalName: (_b = args.profile) === null || _b === void 0 ? void 0 : _b.legalName,
            addressLine1: (((_c = args.profile) === null || _c === void 0 ? void 0 : _c.addressLine1) || '') + (((_d = args.profile) === null || _d === void 0 ? void 0 : _d.addressLine2) || ''),
            addressPostalCode: (_e = args.profile) === null || _e === void 0 ? void 0 : _e.addressPostalCode,
            contactPersonPhone: (_f = args.profile) === null || _f === void 0 ? void 0 : _f.contactPersonPhone,
            contactPersonEmail: (_g = args.profile) === null || _g === void 0 ? void 0 : _g.contactPersonEmail,
            gstin: (_h = args.profile) === null || _h === void 0 ? void 0 : _h.gstin,
            fssaiNumber: (_j = args.profile) === null || _j === void 0 ? void 0 : _j.fssaiNumber,
            licenseNumber: (_k = args.profile) === null || _k === void 0 ? void 0 : _k.licenseNumber,
            bankAccountNo: (_l = args.profile) === null || _l === void 0 ? void 0 : _l.bankAccountNo,
            bankName: (_m = args.profile) === null || _m === void 0 ? void 0 : _m.bankName,
            bankIFSC: (_o = args.profile) === null || _o === void 0 ? void 0 : _o.bankIFSC,
            bankAccountType: (_p = args.profile) === null || _p === void 0 ? void 0 : _p.bankAccountType,
            bankUPI: (_q = args.profile) === null || _q === void 0 ? void 0 : _q.bankUPI,
            additionalInputFieldTitle: (_r = args.profile) === null || _r === void 0 ? void 0 : _r.additionalInputFieldTitle,
            additionalInputFieldValue: (_s = args.profile) === null || _s === void 0 ? void 0 : _s.additionalInputFieldValue,
            additionalDateFieldValue: (_t = args.profile) === null || _t === void 0 ? void 0 : _t.additionalDateFieldValue,
            additionalDateFieldTitle: (_u = args.profile) === null || _u === void 0 ? void 0 : _u.additionalDateFieldTitle,
            signatureLink: ((_v = args.signature) === null || _v === void 0 ? void 0 : _v.imageBase64) || '',
            logoLink: ((_w = args.profile) === null || _w === void 0 ? void 0 : _w.pSetLogoPrintStatus) ? ((_x = args.logo) === null || _x === void 0 ? void 0 : _x.imageBase64) || '' : '',
        };
        billPrint.bill = {
            partyData: {
                ledgerCredit: (_y = args.party) === null || _y === void 0 ? void 0 : _y.credit,
                profileData: {
                    legalName: (_z = args.party) === null || _z === void 0 ? void 0 : _z.businessName,
                    gstNumber: (_0 = args.party) === null || _0 === void 0 ? void 0 : _0.gstin,
                    contactPersonName: (_1 = args.party) === null || _1 === void 0 ? void 0 : _1.name,
                    contactPersonPhone: (_2 = args.party) === null || _2 === void 0 ? void 0 : _2.phone,
                    contactPersonEmail: (_3 = args.party) === null || _3 === void 0 ? void 0 : _3.email,
                    addressLine1: (_4 = args.party) === null || _4 === void 0 ? void 0 : _4.billingAddress,
                    addressState: (_5 = args.party) === null || _5 === void 0 ? void 0 : _5.billingProvience,
                    addressPincode: (_6 = args.party) === null || _6 === void 0 ? void 0 : _6.billingPostalCode,
                    addressOneLine1: (_7 = args.party) === null || _7 === void 0 ? void 0 : _7.deliveryAddress,
                    addressOneState: (_8 = args.party) === null || _8 === void 0 ? void 0 : _8.deliveryProvience,
                    addressOnePincode: (_9 = args.party) === null || _9 === void 0 ? void 0 : _9.deliveryPostalCode,
                    udf1T: '',
                    udf1V: '',
                    udf2T: '',
                    udf2V: '',
                    udf3T: '',
                    udf3V: '',
                    additionalInputFieldTitle: '',
                    additionalInputFieldValue: '',
                    additionalDateFieldValue: '',
                    additionalDateFieldTitle: '',
                },
            },
            isPartyDeleted: args.isPartyDeleted,
            billItems: (_10 = args.estimate.billItems) === null || _10 === void 0 ? void 0 : _10.map(function (billItem) {
                var _a, _b, _c, _d, _e, _f, _g;
                return {
                    itemDes: _this.newLineToBr((billItem === null || billItem === void 0 ? void 0 : billItem.billNote) || ''),
                    itemSrl: '',
                    hsn: ((_a = billItem.item) === null || _a === void 0 ? void 0 : _a.hsn) || '',
                    price: _this.capFractionsToTwo(billItem.price),
                    basePrice: _this.capFractionsToTwo(billItem.basePrice),
                    incTax: billItem.incTax,
                    taxPercentage: _this.capFractionsToTwo(billItem.taxPercentage),
                    cessPercentage: _this.capFractionsToTwo(billItem.cessPercentage),
                    unit: billItem.unit,
                    quantity: billItem.quantity,
                    discountPercentage: _this.capFractionsToTwo(billItem.discountPercentage),
                    discount: _this.capFractionsToTwo(billItem.discount),
                    unitGstAmount: _this.capFractionsToTwo(billItem.unitGstAmount),
                    unitCessAmount: _this.capFractionsToTwo(billItem.unitCessAmount),
                    unitDiscountAmount: _this.capFractionsToTwo(billItem.unitDiscountAmount),
                    unitTaxAmount: _this.capFractionsToTwo(billItem.unitTaxAmount),
                    itemTotalGstAmount: _this.capFractionsToTwo(billItem.itemTotalGstAmount),
                    itemTotalCessAmount: _this.capFractionsToTwo(billItem.itemTotalCessAmount),
                    itemTotalTaxAmount: _this.capFractionsToTwo(billItem.itemTotalTaxAmount),
                    effectiveTaxPercentage: _this.capFractionsToTwo(billItem.effectiveTaxPercentage),
                    billCashDiscountPercentage: _this.capFractionsToTwo(billItem.billCashDiscountPercentage),
                    total: _this.capFractionsToTwo(billItem.wcdTotal),
                    createdStamp: (_b = billItem.item) === null || _b === void 0 ? void 0 : _b.createdStamp,
                    item: {
                        itemName: (_c = billItem.item) === null || _c === void 0 ? void 0 : _c.itemName,
                        hsn: (_d = billItem.item) === null || _d === void 0 ? void 0 : _d.hsn,
                        mrp: (_e = billItem.item) === null || _e === void 0 ? void 0 : _e.mrp,
                        expiryDate: (_f = billItem.item) === null || _f === void 0 ? void 0 : _f.expiryDate,
                    },
                    additionalInputFieldValue: '',
                    additionalDateFieldValue: 0,
                    additionalInputFieldTitle: '',
                    additionalDateFieldTitle: '',
                    mrp: ((_g = billItem.item) === null || _g === void 0 ? void 0 : _g.mrp) || null,
                    variant: '',
                };
            }),
            roundOffValue: this.capFractionsToTwo(args.estimate.roundOffValue) || 0,
            billNo: args.estimate.billNo,
            deliveryState: args.estimate.party.deliveryProvience,
            billDateStamp: args.estimate.billDateStamp,
            dueDateStamp: args.estimate.dueDateStamp,
            purOrderNo: (_11 = args.estimate.transportDetail) === null || _11 === void 0 ? void 0 : _11.purOrderNo,
            eWayBillNo: (_12 = args.estimate.transportDetail) === null || _12 === void 0 ? void 0 : _12.eWayBillNo,
            additionalAmount: this.capFractionsToTwo((_13 = args.estimate.transportDetail) === null || _13 === void 0 ? void 0 : _13.additionalAmount),
            discountAmount: this.capFractionsToTwo(args.estimate.discountAmount),
            gstAmount: this.capFractionsToTwo(args.estimate.gstAmount),
            cessAmount: this.capFractionsToTwo(args.estimate.cessAmount),
            totalAmount: this.capFractionsToTwo(args.estimate.totalAmount),
            createdStamp: args.estimate.createdStamp,
            transporterName: (_14 = args.estimate.transportDetail) === null || _14 === void 0 ? void 0 : _14.transporterName,
            vehicleNumber: (_15 = args.estimate.transportDetail) === null || _15 === void 0 ? void 0 : _15.vehicleNumber,
            deliveryDate: (_16 = args.estimate.transportDetail) === null || _16 === void 0 ? void 0 : _16.deliveryDate,
            color: '#3498db',
            eWayBillDate: (_17 = args.estimate.transportDetail) === null || _17 === void 0 ? void 0 : _17.eWayBillDate,
            note: args.estimate.note,
            billTitle: '',
            tnc: ((_18 = args.profile) === null || _18 === void 0 ? void 0 : _18.pSetTermsAndConditions) || null,
            cashDiscount: this.capFractionsToTwo(args.estimate.cashDiscount),
            cashDiscountPercentage: this.capFractionsToTwo(args.estimate.cashDiscountPercentage),
            challanNo: ((_19 = args.estimate.transportDetail) === null || _19 === void 0 ? void 0 : _19.challanNo) || null,
            totalSaving: this.capFractionsToTwo(args.estimate.totalSaving),
            subTotalAmount: this.capFractionsToTwo(args.estimate.subTotalAmount),
            partyPreviousBalance: this.capFractionsToTwo(args.estimate.partyPreviousBalance) || null,
            serviceChargePercentage: this.capFractionsToTwo(args.estimate.serviceChargePercentage),
            serviceChargeAmount: this.capFractionsToTwo(args.estimate.serviceChargeAmount),
            deliveryProvience: (_20 = args === null || args === void 0 ? void 0 : args.estimate) === null || _20 === void 0 ? void 0 : _20.deliveryProvience,
        };
        if ((_21 = args.secondaryParty) === null || _21 === void 0 ? void 0 : _21._localUUID) {
            billPrint.bill.secondaryPartyData = {
                ledgerCredit: (_22 = args.secondaryParty) === null || _22 === void 0 ? void 0 : _22.credit,
                profileData: {
                    legalName: (_23 = args.secondaryParty) === null || _23 === void 0 ? void 0 : _23.businessName,
                    gstNumber: (_24 = args.secondaryParty) === null || _24 === void 0 ? void 0 : _24.gstin,
                    contactPersonName: (_25 = args.secondaryParty) === null || _25 === void 0 ? void 0 : _25.name,
                    contactPersonPhone: (_26 = args.secondaryParty) === null || _26 === void 0 ? void 0 : _26.phone,
                    contactPersonEmail: (_27 = args.secondaryParty) === null || _27 === void 0 ? void 0 : _27.email,
                    addressLine1: (_28 = args.secondaryParty) === null || _28 === void 0 ? void 0 : _28.billingAddress,
                    addressState: (_29 = args.secondaryParty) === null || _29 === void 0 ? void 0 : _29.billingProvience,
                    addressPincode: (_30 = args.secondaryParty) === null || _30 === void 0 ? void 0 : _30.billingPostalCode,
                    addressOneLine1: (_31 = args.secondaryParty) === null || _31 === void 0 ? void 0 : _31.deliveryAddress,
                    addressOneState: (_32 = args.secondaryParty) === null || _32 === void 0 ? void 0 : _32.deliveryProvience,
                    addressOnePincode: (_33 = args.secondaryParty) === null || _33 === void 0 ? void 0 : _33.deliveryPostalCode,
                    udf1T: '',
                    udf1V: '',
                    udf2T: '',
                    udf2V: '',
                    udf3T: '',
                    udf3V: '',
                    additionalInputFieldTitle: '',
                    additionalInputFieldValue: '',
                    additionalDateFieldValue: '',
                    additionalDateFieldTitle: '',
                },
            };
        }
        // set settings - TODO All
        billPrint.settings = {
            showSignatureNotRequiredText: false,
            showSignatureInTemplates: true,
            makeItemNameBold: false,
            showPreviousAmount: false,
            showTaxAndDiscAmt: (_34 = args.profile) === null || _34 === void 0 ? void 0 : _34.pSetTaxMrpHsnPrintStatus,
            billSavingsAmount: true,
        };
        // set extra - TODO All
        billPrint.extra = {
            thermalTncTemplate: ((_35 = args.profile) === null || _35 === void 0 ? void 0 : _35.pSetTermsAndConditions) || null,
        };
        billPrint.qrPayLink = '';
        billPrint.addressTo = '';
        billPrint.pageWidth = null;
        if (((_36 = args.profile) === null || _36 === void 0 ? void 0 : _36.bankUPI) && ((_37 = args.profile) === null || _37 === void 0 ? void 0 : _37.pSetQrPrintStatus)) {
            var receivedAmount = 0;
            var balancedAmount = 0;
            balancedAmount = ((_38 = args.estimate) === null || _38 === void 0 ? void 0 : _38.totalAmount) - receivedAmount;
            if (((_39 = args.profile) === null || _39 === void 0 ? void 0 : _39.pSetAlwaysQrPrintStatus) || (Number(balancedAmount) || 0) > 0) {
                if (balancedAmount == 0) {
                    balancedAmount = (_40 = args.estimate) === null || _40 === void 0 ? void 0 : _40.totalAmount;
                }
                billPrint.qrPayLink = "upi://pay?pa=".concat((_41 = args.profile) === null || _41 === void 0 ? void 0 : _41.bankUPI, "&pn=").concat((_42 = args.profile) === null || _42 === void 0 ? void 0 : _42.legalName, "&am=").concat(balancedAmount, "&cu=INR&tn=").concat(args.estimate.billNo);
            }
        }
        return billPrint;
    };
    HtmlTemplates.getBillTemplateHtml = function (data, billPref) {
        return __awaiter(this, void 0, void 0, function () {
            var templateObj, error_2, error_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 6, , 7]);
                        templateObj = {
                            temp1MoneyIn: function (x) { return new temp1_money_in_1.Temp1MoneyIn(x).main(); },
                            temp1MoneyOut: function (x) { return new temp1_money_out_1.Temp1MoneyOut(x).main(); },
                            temp1Purchase: function (x) { return new temp1_purchase_1.Temp1Purchase(x).main(); },
                            temp1Expense: function (x) { return new temp1_expense_1.Temp1Expense(x).main(); },
                            temp1Estimate: function (x) { return new temp1_estimate_1.Temp1Estimate(x).main(); },
                            temp6: function (x) { return new temp1_2.Temp1(x).main(); },
                            temp2: function (x) { return new temp2_1.Temp2(x).main(); },
                            temp3: function (x) { return new temp3_1.Temp3(x).main(); },
                            temp9: function (x) { return new modern_A5_1.ModernA5(x).main(); },
                            temp10: function (x) { return new gst1_A5_1.Gst1A5(x).main(); },
                            temp11: function (x) { return new gst2_A5_1.Gst2A5(x).main(); },
                            temp12: function (x) { return new tallyHSN_1.TallyHSN(x).main(); },
                            temp13: function (x) { return new tallyGST_1.TallyGST(x).main(); },
                            temp14: function (x) { return new letter_head_1.LetterHead(x).main(); },
                            temp17: function (x) { return new temp17_1.Temp17(x).main(); },
                            temp17MoneyIn: function (x) { return new temp17_money_in_1.Temp17MoneyIn(x).main(); },
                            temp17MoneyOut: function (x) { return new temp17_money_out_1.Temp17MoneyOut(x).main(); },
                            temp17Purchase: function (x) { return new temp17_purchase_1.Temp17Purchase(x).main(); },
                            temp17Expense: function (x) { return new temp17_expense_1.Temp17Expense(x).main(); },
                            temp17Estimate: function (x) { return new temp17_estimate_1.Temp17Estimate(x).main(); },
                            temp18: function (x) { return new temp18_1.Temp18(x).main(); },
                            temp18MoneyIn: function (x) { return new temp18_money_in_1.Temp18MoneyIn(x).main(); },
                            temp18MoneyOut: function (x) { return new temp18_money_out_1.Temp18MoneyOut(x).main(); },
                            temp18Purchase: function (x) { return new temp18_purchase_1.Temp18Purchase(x).main(); },
                            temp18Expense: function (x) { return new temp18_expense_1.Temp18Expense(x).main(); },
                            temp18Estimate: function (x) { return new temp18_estimate_1.Temp18Estimate(x).main(); },
                            temp19: function (x) { return new temp19_1.Temp19(x).main(); },
                            temp19MoneyIn: function (x) { return new temp19_money_in_1.Temp19MoneyIn(x).main(); },
                            temp19MoneyOut: function (x) { return new temp19_money_out_1.Temp19MoneyOut(x).main(); },
                            temp19Purchase: function (x) { return new temp19_purchase_1.Temp19Purchase(x).main(); },
                            temp19Expense: function (x) { return new temp19_expense_1.Temp19Expense(x).main(); },
                            Temp19Estimate: function (x) { return new temp19_estimate_1.Temp19Estimate(x).main(); },
                            temp20: function (x) { return new temp20_1.Temp20(x).main(); },
                            temp20MoneyIn: function (x) { return new temp20_money_in_1.Temp20MoneyIn(x).main(); },
                            temp20MoneyOut: function (x) { return new temp20_money_out_1.Temp20MoneyOut(x).main(); },
                            temp20Purchase: function (x) { return new temp20_purchase_1.Temp20Purchase(x).main(); },
                            temp20Expense: function (x) { return new temp20_expense_1.Temp20Expense(x).main(); },
                            temp20Estimate: function (x) { return new temp20_estimate_1.Temp20Estimate(x).main(); },
                        };
                        if (!templateObj[billPref]) return [3 /*break*/, 4];
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, templateObj[billPref](data)];
                    case 2: return [2 /*return*/, _a.sent()];
                    case 3:
                        error_2 = _a.sent();
                        console.error("catch error => ", error_2);
                        return [3 /*break*/, 4];
                    case 4: return [4 /*yield*/, templateObj.temp6(data)];
                    case 5: return [2 /*return*/, _a.sent()];
                    case 6:
                        error_3 = _a.sent();
                        return [3 /*break*/, 7];
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    HtmlTemplates.capFractionsToTwo = function (fractionNumber) {
        return Math.round((Number(fractionNumber) || 0) * 100) / 100;
    };
    return HtmlTemplates;
}());
exports.HtmlTemplates = HtmlTemplates;
