// @makarandkate/ezo-connect-wa/src/taggtele.ts
import { Utils } from "./utils";
import request from 'request';

export class Taggtele {

    /**
     * 
     * @param args 
     * @returns  
     */

    public static async telephonicCall(args: {
        phone: string,
        voiceFile: string,
        serviceNumber: string,
        retryAttempt: number,
        retryDuration: number,
    }) {
        args.phone = Utils.sanitizePhone(args.phone);
        
        const reqOptions = {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            url: "https://ezobooks.in/connect/api/taggtele/telephonicCall",
            body: JSON.stringify({
                ...args
            })
        }
        return new Promise((resolve) => {
            request(reqOptions, (error: any, response: any) => {
                if (error) { 
                    console.error(`---------------------------------`)
                    console.error(`EZO-WA:taggtele:35---------------`)
                    console.error(`---------------------------------`)
                    console.error(error); 
                    console.error(`---------------------------------`)
                    resolve(error) 
                };
                resolve(response.body);
            });
        });
    }

}
