// @makarandkate/ezo-connect-wa/src/whatsapp.ts
import { Utils } from "./utils";
import request from 'request';
export enum WhatsappAccountType{
    sales='sales',
    support='support',
    ld='ld'
}
export enum WhatsappVendor{
    sobot='sobot',
    interact='interact'
}
export class WhatsApp {

    /**
     * Sends template message
     * @param args 
     * @returns  
     */
    public static async sendTransactionalMessage(args: {
        vendor?:WhatsappVendor,
        accountType:WhatsappAccountType,
        phone: string,
        templateId: string,
        headerValues: string[],
        bodyValues: string[],
        buttonValues: {
            [key: string]: string[],
        }
    }) {
        args.phone = Utils.sanitizePhone(args.phone);
        let finalBodyValues=[];
        for(let i=0;i<args.bodyValues.length;i++){
            finalBodyValues.push((args.bodyValues[i]+'').replace(/\t/g,'').replace(/\n/g,'').replace(/\r/g,'').replace(/  /,'').replace(/  /,'').replace(/  /,''))
        }
        const reqOptions = {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            url: "https://ezobooks.in/connect/api/message/whatsapp/transactional",
            body: JSON.stringify({
                ...args
            })
        }
        return new Promise((resolve) => {
            request(reqOptions, (error: any, response: any) => {
                if (error) { 
                    console.error(`---------------------------------`)
                    console.error(`EZO-WA:whatsapp:44---------------`)
                    console.error(`---------------------------------`)
                    console.error(error); 
                    console.error(`---------------------------------`)

                    resolve(error) 
                };
                resolve(response.body);
            });
        });
    }

    public static async sendTransactionalMessageRotatory(args: {
        vendor?:WhatsappVendor,
        accountType:WhatsappAccountType,
        phone: string,
        templateIds: string[],
        headerValues: string[],
        bodyValues: string[],
        buttonValues: {
            [key: string]: string[],
        }
    }) {
        args.phone = Utils.sanitizePhone(args.phone);
        let finalBodyValues=[];
        for(let i=0;i<args.bodyValues.length;i++){
            finalBodyValues.push((args.bodyValues[i]+'').replace(/\t/g,'').replace(/\n/g,'').replace(/\r/g,'').replace(/  /,'').replace(/  /,'').replace(/  /,''))
        }
        const reqOptions = {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            url: "https://ezobooks.in/connect/api/message/whatsapp/transactional",
            body: JSON.stringify({
                ...args,
                templateId:args.templateIds[Math.floor(Math.random() * args.templateIds.length)]
            })
        }
        return new Promise((resolve) => {
            request(reqOptions, (error: any, response: any) => {
                if (error) { 
                    console.error(`---------------------------------`)
                    console.error(`EZO-WA:whatsapp:86---------------`)
                    console.error(`---------------------------------`)
                    console.error(error); 
                    console.error(`---------------------------------`)
                    resolve(error) 
                };
                resolve(response.body);
            });
        });
    }

    public static async sendRotTxnMsg(args:{
        vendor?:WhatsappVendor,
        accountType:WhatsappAccountType,
        phone: string,
        eventId: string,
        headerValues: string[],
        bodyValues: string[],
        buttonValues: {
            [key: string]: string[],
        }
    }){
        args.phone = Utils.sanitizePhone(args.phone);
        let finalBodyValues=[];
        for(let i=0;i<args.bodyValues.length;i++){
            finalBodyValues.push((args.bodyValues[i]+'').replace(/\t/g,'').replace(/\n/g,'').replace(/\r/g,'').replace(/  /,'').replace(/  /,'').replace(/  /,''))
        }
        const reqOptions = {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            url: "https://ezobooks.in/connect/api/message/whatsapp/events",
            body: JSON.stringify({
                ...args
            })
        }
        return new Promise((resolve) => {
            request(reqOptions, (error: any, response: any) => {
                if (error) { 
                    console.error(`---------------------------------`)
                    console.error(`EZO-WA:whatsapp:126---------------`)
                    console.error(`---------------------------------`)
                    console.error(error); 
                    console.error(`---------------------------------`)

                    resolve(error) 
                };
                resolve(response.body);
            });
        });
    }
}
