// @makarandkate/ezo-connect-wa/lib/server.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const express_1 = __importDefault(require("express"));
const expressHbs = __importStar(require("express-handlebars"));
const path_1 = __importDefault(require("path"));
const whatsapp_1 = require("./whatsapp");
const app = express_1.default();
const hbs = expressHbs.create({
    helpers: {},
    defaultLayout: "layout",
    partialsDir: ["src/views/partials"],
    extname: ".hbs",
});
app.engine(".hbs", hbs.engine);
app.set("view engine", ".hbs");
const viewsDir = path_1.default.join(__dirname, "views");
app.set("views", viewsDir);
const staticDir = path_1.default.join(__dirname, "public");
app.use(express_1.default.static(staticDir));
app.get('/', (req, res) => __awaiter(void 0, void 0, void 0, function* () {
    whatsapp_1.WhatsApp.sendTransactionalMessageRotatory({
        accountType: whatsapp_1.WhatsappAccountType.sales,
        vendor: whatsapp_1.WhatsappVendor.sobot,
        phone: '7588763862',
        templateIds: ['z91_machine_demo'],
        bodyValues: ['1'],
        headerValues: ['https://s3.ap-south-1.amazonaws.com/assets.sobot.in/114844591702185/None/videos/template-verification/Z91.mp4'],
        buttonValues: {}
    });
    res.send('Boo');
}));
app.listen(9000, () => {
    console.log('Server started on port: ' + 9000);
});
