// @makarandkate/ezo-connect-wa/lib/sms.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sms = void 0;
const request_1 = __importDefault(require("request"));
const utils_1 = require("./utils");
class Sms {
    static sendTemplateSms(config) {
        var msgVar = {
            var1: config.variables[0],
            var2: config.variables[1],
            var3: config.variables[2],
            var4: config.variables[3],
            var5: config.variables[4],
            var6: config.variables[5],
            var7: config.variables[6],
            var8: config.variables[7],
            var9: config.variables[8],
            var10: config.variables[9]
        };
        for (let k in msgVar) {
            if ((msgVar[k] + '') == 'undefined') {
                delete msgVar[k];
            }
            else {
                msgVar[k] = (msgVar[k] + '').replace(/\t/g, '').replace(/\n/g, '').replace(/\r/g, '').replace(/  /, '').replace(/  /, '').replace(/  /, '');
            }
        }
        let urlParam = ``;
        for (let k in msgVar) {
            urlParam += `&${k}=${msgVar[k]}`;
        }
        request_1.default({
            method: 'GET',
            url: `https://ezobooks.in/connect/api/message/sms/transactional?phone=${utils_1.Utils.sanitizePhone(config.phone)}&flowId=${config.flowId}${urlParam}`
        }, (error, res) => {
            if (error) {
                console.error(`---------------------------------`);
                console.error(`EZO-WA:sms:39---------------`);
                console.error(`---------------------------------`);
                console.error(error);
                console.log('config', JSON.stringify(config));
                console.error(`---------------------------------`);
            }
        });
    }
}
exports.Sms = Sms;
