// @makarandkate/ezo-connect-wa/lib/taggtele.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Taggtele = void 0;
const utils_1 = require("./utils");
const request_1 = __importDefault(require("request"));
class Taggtele {
    /**
     *
     * @param args
     * @returns
     */
    static telephonicCall(args) {
        return __awaiter(this, void 0, void 0, function* () {
            args.phone = utils_1.Utils.sanitizePhone(args.phone);
            const reqOptions = {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                },
                url: "https://ezobooks.in/connect/api/taggtele/telephonicCall",
                body: JSON.stringify(Object.assign({}, args))
            };
            return new Promise((resolve) => {
                request_1.default(reqOptions, (error, response) => {
                    if (error) {
                        console.error(`---------------------------------`);
                        console.error(`EZO-WA:taggtele:35---------------`);
                        console.error(`---------------------------------`);
                        console.error(error);
                        console.error(`---------------------------------`);
                        resolve(error);
                    }
                    ;
                    resolve(response.body);
                });
            });
        });
    }
}
exports.Taggtele = Taggtele;
